{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9dd9d21f",
   "metadata": {},
   "source": [
    "This is the second iteration of the data wrangling process.\n",
    "\n",
    "Below is the summary:\n",
    "\n",
    "\n",
    "In short, the project is to create a model that can inform me which nutrients and vitamins I need to monitor and ensure I am getting enough in order to ensure that I am able to gain 10 lbs by September 20th of 2024.\n",
    "\n",
    "\n",
    "\n",
    "Here is a link to the project proposal: https://docs.google.com/document/d/1cmwdacCTAq5hPO_PyioR-l9hnVlCGy5J5BGr9YYgY7M/edit\n",
    "\n",
    "\n",
    "\n",
    "In this data wrangling, I will establish which genes are most significant in the group of the study that had the highest muscle growth.  Then I will link the genes to their perspective enzyme, an enzyme being the functional unit in the muscles that will catalyze change and development in the muscle.  Using the Enzyme Classification system (EC) I will set a classification to each gene that is applicable.  Then I will create a set of associative (summary web scrapes from journal publishing and wikipedia summaries) dictionaries of nutrient and vitamins to their perspective participation in each enzyme classes.  From there, the assocation factors will be multiplied by the membership and traitsig factors of the experiment data set, creating weighted associations of each vitamin and nutrient in question.\n",
    "\n",
    "The explanatory variable in the model will be the gene membership and trait weights as listed by the experiment; these variables are essentially gene clusters that were mapped to their significance in the experimental success, and then the genes were mapped to their significance into these clusters.  \n",
    "\n",
    "The response variables will be the nutrient and vitamin weighted associations.  The model will return which vitamins and minerals are most heavily associated with the genes most heavily implicated in the success of the muscle gain.  \n",
    "\n",
    "The data that will be used is coming from three sources: \n",
    "\n",
    "1.) Below is an article containing data on which gene expressions are most significant in highest muscle growth in a population that was resistance training\n",
    "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8424535/\n",
    "2.) Below is the ncbi datasets containing a plethora of genetic information and specifically descriptions on the enzyzmes created by genes:\n",
    "https://www.ncbi.nlm.nih.gov/datasets\n",
    "3.) Below is a spreadsheet containing all of the sources that were used to scrape information on the different nutrient and vitamins inspected in this:\n",
    "https://docs.google.com/spreadsheets/d/e/2PACX-1vSPbumnC2FG7QbilXNMXzGAeuT9htHy2MNP_qFUBRLxBGheJai-S9E0ecnMUHnjmHlYo7RxXWVPctMI/pubhtml\n",
    "\n",
    "\n",
    "\n",
    "In the process of data wrangling many edits had to take place, as there were missing values and non essential values as well.\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e20f50c",
   "metadata": {},
   "source": [
    "1.) Step One is the uploading of the first data: the data from the muscle growth experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6cdd8bb1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Table S2: Genes in WGCNA Prediction analysis red module (n = 307), strength of association with change in mid-thigh muscle area as measured by CT, and top LV/pathway association in PLIER</th>\n",
       "      <th>Unnamed: 1</th>\n",
       "      <th>Unnamed: 2</th>\n",
       "      <th>Unnamed: 3</th>\n",
       "      <th>Unnamed: 4</th>\n",
       "      <th>Unnamed: 5</th>\n",
       "      <th>Unnamed: 6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>module</td>\n",
       "      <td>membership</td>\n",
       "      <td>traitsig</td>\n",
       "      <td>ensg_version</td>\n",
       "      <td>gene</td>\n",
       "      <td>LV_number</td>\n",
       "      <td>LV_pathway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>red</td>\n",
       "      <td>0.5158</td>\n",
       "      <td>0.2172</td>\n",
       "      <td>ENSG00000137959.15</td>\n",
       "      <td>IFI44L</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>red</td>\n",
       "      <td>0.5497</td>\n",
       "      <td>0.4713</td>\n",
       "      <td>ENSG00000160710.16</td>\n",
       "      <td>ADAR</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>red</td>\n",
       "      <td>0.7754</td>\n",
       "      <td>0.5447</td>\n",
       "      <td>ENSG00000082512.14</td>\n",
       "      <td>TRAF5</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Table S2: Genes in WGCNA Prediction analysis red module (n = 307), strength of association with change in mid-thigh muscle area as measured by CT, and top LV/pathway association in PLIER  \\\n",
       "0                                                NaN                                                                                                                                           \n",
       "1                                             module                                                                                                                                           \n",
       "2                                                red                                                                                                                                           \n",
       "3                                                red                                                                                                                                           \n",
       "4                                                red                                                                                                                                           \n",
       "\n",
       "   Unnamed: 1 Unnamed: 2          Unnamed: 3 Unnamed: 4 Unnamed: 5  \\\n",
       "0         NaN        NaN                 NaN        NaN        NaN   \n",
       "1  membership   traitsig        ensg_version       gene  LV_number   \n",
       "2      0.5158     0.2172  ENSG00000137959.15     IFI44L          2   \n",
       "3      0.5497     0.4713  ENSG00000160710.16       ADAR          2   \n",
       "4      0.7754     0.5447  ENSG00000082512.14      TRAF5          2   \n",
       "\n",
       "                Unnamed: 6  \n",
       "0                      NaN  \n",
       "1               LV_pathway  \n",
       "2  MetaMEx_AerobicTraining  \n",
       "3  MetaMEx_AerobicTraining  \n",
       "4  MetaMEx_AerobicTraining  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "df= pd.read_excel('Table_S2.xlsx')\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f98d971",
   "metadata": {},
   "source": [
    "Editing the dataframe so it is digestable and inspecting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "34f59577",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>1</th>\n",
       "      <th>module</th>\n",
       "      <th>membership</th>\n",
       "      <th>traitsig</th>\n",
       "      <th>ensg_version</th>\n",
       "      <th>gene</th>\n",
       "      <th>LV_number</th>\n",
       "      <th>LV_pathway</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>module</td>\n",
       "      <td>membership</td>\n",
       "      <td>traitsig</td>\n",
       "      <td>ensg_version</td>\n",
       "      <td>gene</td>\n",
       "      <td>LV_number</td>\n",
       "      <td>LV_pathway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>red</td>\n",
       "      <td>0.5158</td>\n",
       "      <td>0.2172</td>\n",
       "      <td>ENSG00000137959.15</td>\n",
       "      <td>IFI44L</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>red</td>\n",
       "      <td>0.5497</td>\n",
       "      <td>0.4713</td>\n",
       "      <td>ENSG00000160710.16</td>\n",
       "      <td>ADAR</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>red</td>\n",
       "      <td>0.7754</td>\n",
       "      <td>0.5447</td>\n",
       "      <td>ENSG00000082512.14</td>\n",
       "      <td>TRAF5</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "1  module  membership  traitsig        ensg_version    gene  LV_number  \\\n",
       "0     NaN         NaN       NaN                 NaN     NaN        NaN   \n",
       "1  module  membership  traitsig        ensg_version    gene  LV_number   \n",
       "2     red      0.5158    0.2172  ENSG00000137959.15  IFI44L          2   \n",
       "3     red      0.5497    0.4713  ENSG00000160710.16    ADAR          2   \n",
       "4     red      0.7754    0.5447  ENSG00000082512.14   TRAF5          2   \n",
       "\n",
       "1               LV_pathway  \n",
       "0                      NaN  \n",
       "1               LV_pathway  \n",
       "2  MetaMEx_AerobicTraining  \n",
       "3  MetaMEx_AerobicTraining  \n",
       "4  MetaMEx_AerobicTraining  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns = df.iloc[1]\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "62cc9193",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>1</th>\n",
       "      <th>module</th>\n",
       "      <th>membership</th>\n",
       "      <th>traitsig</th>\n",
       "      <th>ensg_version</th>\n",
       "      <th>gene</th>\n",
       "      <th>LV_number</th>\n",
       "      <th>LV_pathway</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>red</td>\n",
       "      <td>0.5158</td>\n",
       "      <td>0.2172</td>\n",
       "      <td>ENSG00000137959.15</td>\n",
       "      <td>IFI44L</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>red</td>\n",
       "      <td>0.5497</td>\n",
       "      <td>0.4713</td>\n",
       "      <td>ENSG00000160710.16</td>\n",
       "      <td>ADAR</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>red</td>\n",
       "      <td>0.7754</td>\n",
       "      <td>0.5447</td>\n",
       "      <td>ENSG00000082512.14</td>\n",
       "      <td>TRAF5</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>red</td>\n",
       "      <td>0.7309</td>\n",
       "      <td>0.4325</td>\n",
       "      <td>ENSG00000163219.11</td>\n",
       "      <td>ARHGAP25</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>red</td>\n",
       "      <td>0.5587</td>\n",
       "      <td>0.4396</td>\n",
       "      <td>ENSG00000153208.16</td>\n",
       "      <td>MERTK</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "1 module membership traitsig        ensg_version      gene LV_number  \\\n",
       "2    red     0.5158   0.2172  ENSG00000137959.15    IFI44L         2   \n",
       "3    red     0.5497   0.4713  ENSG00000160710.16      ADAR         2   \n",
       "4    red     0.7754   0.5447  ENSG00000082512.14     TRAF5         2   \n",
       "5    red     0.7309   0.4325  ENSG00000163219.11  ARHGAP25         2   \n",
       "6    red     0.5587   0.4396  ENSG00000153208.16     MERTK         2   \n",
       "\n",
       "1               LV_pathway  \n",
       "2  MetaMEx_AerobicTraining  \n",
       "3  MetaMEx_AerobicTraining  \n",
       "4  MetaMEx_AerobicTraining  \n",
       "5  MetaMEx_AerobicTraining  \n",
       "6  MetaMEx_AerobicTraining  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.drop([0,1])\n",
    "df.reset_index()\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "01c466d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 307 entries, 2 to 308\n",
      "Data columns (total 7 columns):\n",
      " #   Column        Non-Null Count  Dtype \n",
      "---  ------        --------------  ----- \n",
      " 0   module        307 non-null    object\n",
      " 1   membership    307 non-null    object\n",
      " 2   traitsig      307 non-null    object\n",
      " 3   ensg_version  307 non-null    object\n",
      " 4   gene          307 non-null    object\n",
      " 5   LV_number     97 non-null     object\n",
      " 6   LV_pathway    97 non-null     object\n",
      "dtypes: object(7)\n",
      "memory usage: 16.9+ KB\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>1</th>\n",
       "      <th>module</th>\n",
       "      <th>membership</th>\n",
       "      <th>traitsig</th>\n",
       "      <th>ensg_version</th>\n",
       "      <th>gene</th>\n",
       "      <th>LV_number</th>\n",
       "      <th>LV_pathway</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>307</td>\n",
       "      <td>307.0000</td>\n",
       "      <td>307.0000</td>\n",
       "      <td>307</td>\n",
       "      <td>307</td>\n",
       "      <td>97</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>1</td>\n",
       "      <td>297.0000</td>\n",
       "      <td>293.0000</td>\n",
       "      <td>307</td>\n",
       "      <td>307</td>\n",
       "      <td>15</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>red</td>\n",
       "      <td>0.4352</td>\n",
       "      <td>0.4325</td>\n",
       "      <td>ENSG00000137959.15</td>\n",
       "      <td>IFI44L</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>307</td>\n",
       "      <td>3.0000</td>\n",
       "      <td>3.0000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>41</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "1      module  membership  traitsig        ensg_version    gene  LV_number  \\\n",
       "count     307    307.0000  307.0000                 307     307         97   \n",
       "unique      1    297.0000  293.0000                 307     307         15   \n",
       "top       red      0.4352    0.4325  ENSG00000137959.15  IFI44L          2   \n",
       "freq      307      3.0000    3.0000                   1       1         41   \n",
       "\n",
       "1                    LV_pathway  \n",
       "count                        97  \n",
       "unique                       10  \n",
       "top     MetaMEx_AerobicTraining  \n",
       "freq                         46  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.info()\n",
    "df.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "63a40532",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1\n",
       "module            0\n",
       "membership        0\n",
       "traitsig          0\n",
       "ensg_version      0\n",
       "gene              0\n",
       "LV_number       210\n",
       "LV_pathway      210\n",
       "dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.isna().any()\n",
    "df.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e20f1d5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>1</th>\n",
       "      <th>membership</th>\n",
       "      <th>traitsig</th>\n",
       "      <th>ensg_version</th>\n",
       "      <th>gene</th>\n",
       "      <th>LV_number</th>\n",
       "      <th>LV_pathway</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.5158</td>\n",
       "      <td>0.2172</td>\n",
       "      <td>ENSG00000137959.15</td>\n",
       "      <td>IFI44L</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.5497</td>\n",
       "      <td>0.4713</td>\n",
       "      <td>ENSG00000160710.16</td>\n",
       "      <td>ADAR</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.7754</td>\n",
       "      <td>0.5447</td>\n",
       "      <td>ENSG00000082512.14</td>\n",
       "      <td>TRAF5</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.7309</td>\n",
       "      <td>0.4325</td>\n",
       "      <td>ENSG00000163219.11</td>\n",
       "      <td>ARHGAP25</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.5587</td>\n",
       "      <td>0.4396</td>\n",
       "      <td>ENSG00000153208.16</td>\n",
       "      <td>MERTK</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "1 membership traitsig        ensg_version      gene LV_number  \\\n",
       "2     0.5158   0.2172  ENSG00000137959.15    IFI44L         2   \n",
       "3     0.5497   0.4713  ENSG00000160710.16      ADAR         2   \n",
       "4     0.7754   0.5447  ENSG00000082512.14     TRAF5         2   \n",
       "5     0.7309   0.4325  ENSG00000163219.11  ARHGAP25         2   \n",
       "6     0.5587   0.4396  ENSG00000153208.16     MERTK         2   \n",
       "\n",
       "1               LV_pathway  \n",
       "2  MetaMEx_AerobicTraining  \n",
       "3  MetaMEx_AerobicTraining  \n",
       "4  MetaMEx_AerobicTraining  \n",
       "5  MetaMEx_AerobicTraining  \n",
       "6  MetaMEx_AerobicTraining  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.drop(columns=['module'])\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "030dd7a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>1</th>\n",
       "      <th>membership</th>\n",
       "      <th>traitsig</th>\n",
       "      <th>ensg_version</th>\n",
       "      <th>gene</th>\n",
       "      <th>LV_number</th>\n",
       "      <th>LV_pathway</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.5158</td>\n",
       "      <td>0.2172</td>\n",
       "      <td>ENSG00000137959.15</td>\n",
       "      <td>IFI44L</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.5497</td>\n",
       "      <td>0.4713</td>\n",
       "      <td>ENSG00000160710.16</td>\n",
       "      <td>ADAR</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.7754</td>\n",
       "      <td>0.5447</td>\n",
       "      <td>ENSG00000082512.14</td>\n",
       "      <td>TRAF5</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.7309</td>\n",
       "      <td>0.4325</td>\n",
       "      <td>ENSG00000163219.11</td>\n",
       "      <td>ARHGAP25</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.5587</td>\n",
       "      <td>0.4396</td>\n",
       "      <td>ENSG00000153208.16</td>\n",
       "      <td>MERTK</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "1  membership  traitsig        ensg_version      gene LV_number  \\\n",
       "2      0.5158    0.2172  ENSG00000137959.15    IFI44L         2   \n",
       "3      0.5497    0.4713  ENSG00000160710.16      ADAR         2   \n",
       "4      0.7754    0.5447  ENSG00000082512.14     TRAF5         2   \n",
       "5      0.7309    0.4325  ENSG00000163219.11  ARHGAP25         2   \n",
       "6      0.5587    0.4396  ENSG00000153208.16     MERTK         2   \n",
       "\n",
       "1               LV_pathway  \n",
       "2  MetaMEx_AerobicTraining  \n",
       "3  MetaMEx_AerobicTraining  \n",
       "4  MetaMEx_AerobicTraining  \n",
       "5  MetaMEx_AerobicTraining  \n",
       "6  MetaMEx_AerobicTraining  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.astype({'membership': 'float64'})\n",
    "df = df.astype({'traitsig': 'float64'})\n",
    "df = df.astype({'LV_number': 'category'})\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8d5b7bbc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    307.000000\n",
       "mean       0.569397\n",
       "std        0.117803\n",
       "min        0.209100\n",
       "25%        0.475950\n",
       "50%        0.572400\n",
       "75%        0.666600\n",
       "max        0.838100\n",
       "Name: membership, dtype: float64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['membership'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9a8b365f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    307.000000\n",
       "mean       0.424034\n",
       "std        0.133645\n",
       "min        0.029100\n",
       "25%        0.333600\n",
       "50%        0.434500\n",
       "75%        0.524300\n",
       "max        0.765800\n",
       "Name: traitsig, dtype: float64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['traitsig'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "516cd2cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>1</th>\n",
       "      <th>membership</th>\n",
       "      <th>traitsig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>97.000000</td>\n",
       "      <td>97.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.618047</td>\n",
       "      <td>0.426052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.107244</td>\n",
       "      <td>0.146436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.343900</td>\n",
       "      <td>0.082200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.549700</td>\n",
       "      <td>0.326400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>0.637900</td>\n",
       "      <td>0.447500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.698600</td>\n",
       "      <td>0.526900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>0.838100</td>\n",
       "      <td>0.765800</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "1      membership   traitsig\n",
       "count   97.000000  97.000000\n",
       "mean     0.618047   0.426052\n",
       "std      0.107244   0.146436\n",
       "min      0.343900   0.082200\n",
       "25%      0.549700   0.326400\n",
       "50%      0.637900   0.447500\n",
       "75%      0.698600   0.526900\n",
       "max      0.838100   0.765800"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not_null = df.notnull().all(axis=1)\n",
    "not_null_df = df[not_null]\n",
    "not_null_df.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3344983e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAh8AAAGdCAYAAACyzRGfAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAiUklEQVR4nO3df3BU1f3/8ddKZAmYjQIm2Zh8MlETFBBqQQPxBwEhNXVQQFsqykBHHSlozTCKBNoaOpoorYg2wohFBCWGaRVlhp9hMAGk2IAwRbCAGjQM+SFIsuFHlwHO9w+H/boG0JvsnmXD8zFzZ7znnr33vcdl85qzd8+6jDFGAAAAllwS6QIAAMDFhfABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwKqYSBfwQ6dPn9aBAwcUFxcnl8sV6XIAAMBPYIxRc3OzkpOTdckl55/buODCx4EDB5SamhrpMgAAQCvU1NQoJSXlvH0uuPARFxcn6bviPR5PhKsBAAA/hc/nU2pqauDv+PlccOHjzEctHo+H8AEAQJT5KbdMcMMpAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsiol0AQDgxNpd9RG57tCeiRG5LtAeMfMBAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwqk3ho7i4WC6XS/n5+YE2Y4wKCwuVnJys2NhY5eTkaOfOnW2tEwAAtBOtDh9VVVWaN2+e+vTpE9Q+c+ZMzZo1SyUlJaqqqlJSUpKGDRum5ubmNhcLAACiX6vCx5EjR/TAAw/o9ddf1xVXXBFoN8Zo9uzZmj59ukaNGqXevXtr4cKFOnbsmEpLS0NWNAAAiF6tCh+TJk3SXXfdpaFDhwa1V1dXq66uTrm5uYE2t9utQYMGadOmTWc9l9/vl8/nC9oAAED7FeP0AWVlZfrkk09UVVXV4lhdXZ0kKTExMag9MTFRX3311VnPV1xcrBkzZjgtA7igrN1VH5HrDu2Z+OOdAOAC42jmo6amRk888YTefvttderU6Zz9XC5X0L4xpkXbGQUFBWpqagpsNTU1TkoCAABRxtHMx9atW9XQ0KB+/foF2k6dOqX169erpKREu3fvlvTdDIjX6w30aWhoaDEbcobb7Zbb7W5N7QAAIAo5mvm44447tGPHDm3fvj2w9e/fXw888IC2b9+uq6++WklJSSovLw885sSJE6qsrFR2dnbIiwcAANHH0cxHXFycevfuHdTWpUsXdevWLdCen5+voqIiZWRkKCMjQ0VFRercubPGjBkTuqoBAEDUcnzD6Y+ZMmWKjh8/rokTJ+rw4cPKysrSmjVrFBcXF+pLAQCAKOQyxphIF/F9Pp9P8fHxampqksfjiXQ5wE/Ct13sYayBC5OTv9/8tgsAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsCvlvuwBAe8Sy7kDoMPMBAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrW+QCiGGtPAIhGzHwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArGKRMQBACyxgh3Bi5gMAAFhF+AAAAFYRPgAAgFWOwsfcuXPVp08feTweeTweDRw4UCtXrgwcHz9+vFwuV9A2YMCAkBcNAACil6MbTlNSUvT888/r2muvlSQtXLhQ99xzj7Zt26ZevXpJku68804tWLAg8JiOHTuGsFwAABDtHIWP4cOHB+0/99xzmjt3rjZv3hwIH263W0lJSaGrEAAAtCutvufj1KlTKisr09GjRzVw4MBAe0VFhRISEpSZmalHHnlEDQ0N5z2P3++Xz+cL2gAAQPvlOHzs2LFDl112mdxutyZMmKClS5eqZ8+ekqS8vDwtXrxY69at04svvqiqqioNGTJEfr//nOcrLi5WfHx8YEtNTW39swEAABc8lzHGOHnAiRMn9PXXX6uxsVHvvvuu/v73v6uysjIQQL6vtrZWaWlpKisr06hRo856Pr/fHxROfD6fUlNT1dTUJI/H4/DpAJERqQWZIiWSC0Ex1nawyBic8vl8io+P/0l/vx2vcNqxY8fADaf9+/dXVVWVXn75Zb322mst+nq9XqWlpWnv3r3nPJ/b7Zbb7XZaBgAAiFJtXufDGHPOj1UOHTqkmpoaeb3etl4GAAC0E45mPqZNm6a8vDylpqaqublZZWVlqqio0KpVq3TkyBEVFhbq3nvvldfr1b59+zRt2jR1795dI0eODFf9AAAgyjgKH/X19Ro7dqxqa2sVHx+vPn36aNWqVRo2bJiOHz+uHTt2aNGiRWpsbJTX69XgwYO1ZMkSxcXFhat+AAAQZRyFj/nz55/zWGxsrFavXt3mggAAQPvGb7sAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsMpR+Jg7d6769Okjj8cjj8ejgQMHauXKlYHjxhgVFhYqOTlZsbGxysnJ0c6dO0NeNAAAiF6OwkdKSoqef/55bdmyRVu2bNGQIUN0zz33BALGzJkzNWvWLJWUlKiqqkpJSUkaNmyYmpubw1I8AACIPo7Cx/Dhw/XLX/5SmZmZyszM1HPPPafLLrtMmzdvljFGs2fP1vTp0zVq1Cj17t1bCxcu1LFjx1RaWhqu+gEAQJRp9T0fp06dUllZmY4ePaqBAwequrpadXV1ys3NDfRxu90aNGiQNm3adM7z+P1++Xy+oA0AALRfjsPHjh07dNlll8ntdmvChAlaunSpevbsqbq6OklSYmJiUP/ExMTAsbMpLi5WfHx8YEtNTXVaEgAAiCKOw0ePHj20fft2bd68Wb/73e80btw47dq1K3Dc5XIF9TfGtGj7voKCAjU1NQW2mpoapyUBAIAoEuP0AR07dtS1114rSerfv7+qqqr08ssv6+mnn5Yk1dXVyev1Bvo3NDS0mA35PrfbLbfb7bQMAAAQpdq8zocxRn6/X+np6UpKSlJ5eXng2IkTJ1RZWans7Oy2XgYAALQTjmY+pk2bpry8PKWmpqq5uVllZWWqqKjQqlWr5HK5lJ+fr6KiImVkZCgjI0NFRUXq3LmzxowZE676AQBAlHEUPurr6zV27FjV1tYqPj5effr00apVqzRs2DBJ0pQpU3T8+HFNnDhRhw8fVlZWltasWaO4uLiwFA8AAKKPo/Axf/788x53uVwqLCxUYWFhW2oCAADtGL/tAgAArCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKscLa8OALBr7a76SJcAhBwzHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrHIWP4uJi3XTTTYqLi1NCQoJGjBih3bt3B/UZP368XC5X0DZgwICQFg0AAKKXo/BRWVmpSZMmafPmzSovL9fJkyeVm5uro0ePBvW78847VVtbG9hWrFgR0qIBAED0inHSedWqVUH7CxYsUEJCgrZu3arbb7890O52u5WUlBSaCgEAQLvSpns+mpqaJEldu3YNaq+oqFBCQoIyMzP1yCOPqKGh4Zzn8Pv98vl8QRsAAGi/Wh0+jDGaPHmybr31VvXu3TvQnpeXp8WLF2vdunV68cUXVVVVpSFDhsjv95/1PMXFxYqPjw9sqamprS0JAABEAZcxxrTmgZMmTdLy5cu1ceNGpaSknLNfbW2t0tLSVFZWplGjRrU47vf7g4KJz+dTamqqmpqa5PF4WlMaYN3aXfWRLsGqoT0TI3bti22sLzaRfG2hbXw+n+Lj43/S329H93yc8fjjj2vZsmVav379eYOHJHm9XqWlpWnv3r1nPe52u+V2u1tTBgAAiEKOwocxRo8//riWLl2qiooKpaen/+hjDh06pJqaGnm93lYXCQAA2g9H93xMmjRJb7/9tkpLSxUXF6e6ujrV1dXp+PHjkqQjR47oySef1L/+9S/t27dPFRUVGj58uLp3766RI0eG5QkAAIDo4mjmY+7cuZKknJycoPYFCxZo/Pjx6tChg3bs2KFFixapsbFRXq9XgwcP1pIlSxQXFxeyogEAQPRy/LHL+cTGxmr16tVtKggAALRv/LYLAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArHK0vDoASNLaXfWRLgFAFGPmAwAAWEX4AAAAVhE+AACAVYQPAABgFeEDAABYRfgAAABWET4AAIBVrPMBALhgRGoNmaE9EyNy3YsVMx8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACschQ+iouLddNNNykuLk4JCQkaMWKEdu/eHdTHGKPCwkIlJycrNjZWOTk52rlzZ0iLBgAA0ctR+KisrNSkSZO0efNmlZeX6+TJk8rNzdXRo0cDfWbOnKlZs2appKREVVVVSkpK0rBhw9Tc3Bzy4gEAQPRxGWNMax/8zTffKCEhQZWVlbr99ttljFFycrLy8/P19NNPS5L8fr8SExP1wgsv6NFHH/3Rc/p8PsXHx6upqUkej6e1pQFWReqXOAGEBr9q23ZO/n636Z6PpqYmSVLXrl0lSdXV1aqrq1Nubm6gj9vt1qBBg7Rp06a2XAoAALQTMa19oDFGkydP1q233qrevXtLkurq6iRJiYnBCTIxMVFfffXVWc/j9/vl9/sD+z6fr7UlAQCAKNDqmY/HHntM//nPf/TOO++0OOZyuYL2jTEt2s4oLi5WfHx8YEtNTW1tSQAAIAq0Knw8/vjjWrZsmT788EOlpKQE2pOSkiT9/xmQMxoaGlrMhpxRUFCgpqamwFZTU9OakgAAQJRwFD6MMXrsscf03nvvad26dUpPTw86np6erqSkJJWXlwfaTpw4ocrKSmVnZ5/1nG63Wx6PJ2gDAADtl6N7PiZNmqTS0lJ98MEHiouLC8xwxMfHKzY2Vi6XS/n5+SoqKlJGRoYyMjJUVFSkzp07a8yYMWF5AgAAILo4Ch9z586VJOXk5AS1L1iwQOPHj5ckTZkyRcePH9fEiRN1+PBhZWVlac2aNYqLiwtJwQAAILq1aZ2PcGCdD0Qj1vkAohvrfLSdtXU+AAAAnCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqx+Fj/fr1Gj58uJKTk+VyufT+++8HHR8/frxcLlfQNmDAgFDVCwAAopzj8HH06FH17dtXJSUl5+xz5513qra2NrCtWLGiTUUCAID2I8bpA/Ly8pSXl3fePm63W0lJSa0uCgAAtF9hueejoqJCCQkJyszM1COPPKKGhoZwXAYAAEQhxzMfPyYvL0+/+tWvlJaWpurqav3xj3/UkCFDtHXrVrnd7hb9/X6//H5/YN/n84W6JAAAcAEJefgYPXp04L979+6t/v37Ky0tTcuXL9eoUaNa9C8uLtaMGTNCXQYAIAS6H1gXtnMfTB4StnPjwhb2r9p6vV6lpaVp7969Zz1eUFCgpqamwFZTUxPukgAAQASFfObjhw4dOqSamhp5vd6zHne73Wf9OAYAALRPjsPHkSNH9Pnnnwf2q6urtX37dnXt2lVdu3ZVYWGh7r33Xnm9Xu3bt0/Tpk1T9+7dNXLkyJAWDgAAopPj8LFlyxYNHjw4sD958mRJ0rhx4zR37lzt2LFDixYtUmNjo7xerwYPHqwlS5YoLi4udFUDAICo5Th85OTkyBhzzuOrV69uU0EAAKB947ddAACAVYQPAABgVdi/7QIAF7pwrmUhsZ4F8EPMfAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwinU+0K6s3VUf6RIAAD+CmQ8AAGAV4QMAAFhF+AAAAFYRPgAAgFWEDwAAYBXhAwAAWEX4AAAAVhE+AACAVSwyBgBRrPuBdZEuAXCMmQ8AAGAV4QMAAFhF+AAAAFYRPgAAgFWEDwAAYBXhAwAAWEX4AAAAVrHOBwDgord2V31Erju0Z2JErhtpzHwAAACrCB8AAMAqwgcAALDKcfhYv369hg8fruTkZLlcLr3//vtBx40xKiwsVHJysmJjY5WTk6OdO3eGql4AABDlHIePo0ePqm/fviopKTnr8ZkzZ2rWrFkqKSlRVVWVkpKSNGzYMDU3N7e5WAAAEP0cf9slLy9PeXl5Zz1mjNHs2bM1ffp0jRo1SpK0cOFCJSYmqrS0VI8++mjbqgUAAFEvpPd8VFdXq66uTrm5uYE2t9utQYMGadOmTWd9jN/vl8/nC9oAAED7FdLwUVdXJ0lKTAz+3nJiYmLg2A8VFxcrPj4+sKWmpoayJAAAcIEJy7ddXC5X0L4xpkXbGQUFBWpqagpsNTU14SgJAABcIEK6wmlSUpKk72ZAvF5voL2hoaHFbMgZbrdbbrc7lGUAAIALWEhnPtLT05WUlKTy8vJA24kTJ1RZWans7OxQXgoAAEQpxzMfR44c0eeffx7Yr66u1vbt29W1a1f93//9n/Lz81VUVKSMjAxlZGSoqKhInTt31pgxY0JaOAAAiE6Ow8eWLVs0ePDgwP7kyZMlSePGjdObb76pKVOm6Pjx45o4caIOHz6srKwsrVmzRnFxcaGrGgAARC3H4SMnJ0fGmHMed7lcKiwsVGFhYVvqAgAA7RS/7QIAAKwifAAAAKtC+lVbABeu7gfWhfX8B5OHhPX8aH94TV68mPkAAABWET4AAIBVhA8AAGAV4QMAAFhF+AAAAFYRPgAAgFWEDwAAYBXrfAC44IV7PQgAdjHzAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwKibSBaB9WrurPtIlALjIdT+wLqznP5g8JKznb8+Y+QAAAFYRPgAAgFWEDwAAYFXIw0dhYaFcLlfQlpSUFOrLAACAKBWWG0579eqltWvXBvY7dOgQjssAAIAoFJbwERMTw2wHAAA4q7Dc87F3714lJycrPT1dv/nNb/Tll1+es6/f75fP5wvaAABA+xXymY+srCwtWrRImZmZqq+v17PPPqvs7Gzt3LlT3bp1a9G/uLhYM2bMCHUZACwL95oKQHsUqTWRhvZMjMh1zwj5zEdeXp7uvfde3XDDDRo6dKiWL18uSVq4cOFZ+xcUFKipqSmw1dTUhLokAABwAQn7CqddunTRDTfcoL179571uNvtltvtDncZAADgAhH2dT78fr8+++wzeb3ecF8KAABEgZCHjyeffFKVlZWqrq7Wxx9/rPvuu08+n0/jxo0L9aUAAEAUCvnHLvv379f999+vgwcP6sorr9SAAQO0efNmpaWlhfpSAAAgCoU8fJSVlYX6lAAAoB3ht10AAIBVhA8AAGBV2L9qC7Qn4V5I62DykLCeH5HBAmxAMGY+AACAVYQPAABgFeEDAABYRfgAAABWET4AAIBVhA8AAGAV4QMAAFjFOh/t3Npd9ZEuAQ6wHgQQPcL577W9r/nDzAcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAqotukbFILbo1tGdiRK57sWGRLgDtQdjfy3reH97z/whmPgAAgFWEDwAAYBXhAwAAWEX4AAAAVhE+AACAVYQPAABgFeEDAABYddGt8xEpkVpf5ELEWhwAcHFj5gMAAFhF+AAAAFYRPgAAgFVhCx9z5sxRenq6OnXqpH79+mnDhg3huhQAAIgiYQkfS5YsUX5+vqZPn65t27bptttuU15enr7++utwXA4AAESRsISPWbNm6aGHHtLDDz+s66+/XrNnz1Zqaqrmzp0bjssBAIAoEvKv2p44cUJbt27V1KlTg9pzc3O1adOmFv39fr/8fn9gv6mpSZLk8/lCXZok6eiR5rCcFz9dp6PHIl0CAFzUwvE39sw5jTE/2jfk4ePgwYM6deqUEhMTg9oTExNVV1fXon9xcbFmzJjRoj01NTXUpQEAAEnSw2E7c3Nzs+Lj48/bJ2yLjLlcrqB9Y0yLNkkqKCjQ5MmTA/unT5/Wt99+q27dup21/4/x+XxKTU1VTU2NPB6P88LbOcbn/Bif82N8zo2xOT/G5/zaw/gYY9Tc3Kzk5OQf7Rvy8NG9e3d16NChxSxHQ0NDi9kQSXK73XK73UFtl19+eZvr8Hg8Ufs/0AbG5/wYn/NjfM6NsTk/xuf8on18fmzG44yQ33DasWNH9evXT+Xl5UHt5eXlys7ODvXlAABAlAnLxy6TJ0/W2LFj1b9/fw0cOFDz5s3T119/rQkTJoTjcgAAIIqEJXyMHj1ahw4d0p///GfV1taqd+/eWrFihdLS0sJxuSBut1vPPPNMi49y8B3G5/wYn/NjfM6NsTk/xuf8LrbxcZmf8p0YAACAEOG3XQAAgFWEDwAAYBXhAwAAWEX4AAAAVkVl+JgzZ47S09PVqVMn9evXTxs2bDhn3/fee0/Dhg3TlVdeKY/Ho4EDB2r16tUWq7XPyfhs3LhRt9xyi7p166bY2Fhdd911eumllyxWa5+T8fm+jz76SDExMfrZz34W3gIjyMnYVFRUyOVytdj++9//WqzYLqevHb/fr+nTpystLU1ut1vXXHON3njjDUvV2udkfMaPH3/W10+vXr0sVmyP09fO4sWL1bdvX3Xu3Fler1e//e1vdejQIUvVWmCiTFlZmbn00kvN66+/bnbt2mWeeOIJ06VLF/PVV1+dtf8TTzxhXnjhBfPvf//b7NmzxxQUFJhLL73UfPLJJ5Yrt8Pp+HzyySemtLTUfPrpp6a6utq89dZbpnPnzua1116zXLkdTsfnjMbGRnP11Veb3Nxc07dvXzvFWuZ0bD788EMjyezevdvU1tYGtpMnT1qu3I7WvHbuvvtuk5WVZcrLy011dbX5+OOPzUcffWSxanucjk9jY2PQ66ampsZ07drVPPPMM3YLt8Dp2GzYsMFccskl5uWXXzZffvml2bBhg+nVq5cZMWKE5crDJ+rCx80332wmTJgQ1HbdddeZqVOn/uRz9OzZ08yYMSPUpV0QQjE+I0eONA8++GCoS7sgtHZ8Ro8ebf7whz+YZ555pt2GD6djcyZ8HD582EJ1ked0fFauXGni4+PNoUOHbJQXcW1971m6dKlxuVxm37594SgvopyOzV/+8hdz9dVXB7W98sorJiUlJWw12hZVH7ucOHFCW7duVW5ublB7bm6uNm3a9JPOcfr0aTU3N6tr167hKDGiQjE+27Zt06ZNmzRo0KBwlBhRrR2fBQsW6IsvvtAzzzwT7hIjpi2vnRtvvFFer1d33HGHPvzww3CWGTGtGZ9ly5apf//+mjlzpq666iplZmbqySef1PHjx22UbFUo3nvmz5+voUOHWlmM0qbWjE12drb279+vFStWyBij+vp6/fOf/9Rdd91lo2QrwvartuFw8OBBnTp1qsUP1CUmJrb4IbtzefHFF3X06FH9+te/DkeJEdWW8UlJSdE333yjkydPqrCwUA8/HL6fW46U1ozP3r17NXXqVG3YsEExMVH1z8WR1oyN1+vVvHnz1K9fP/n9fr311lu64447VFFRodtvv91G2da0Zny+/PJLbdy4UZ06ddLSpUt18OBBTZw4Ud9++227u++jre/NtbW1WrlypUpLS8NVYsS0Zmyys7O1ePFijR49Wv/73/908uRJ3X333frb3/5mo2QrovLd1OVyBe0bY1q0nc0777yjwsJCffDBB0pISAhXeRHXmvHZsGGDjhw5os2bN2vq1Km69tprdf/994ezzIj5qeNz6tQpjRkzRjNmzFBmZqat8iLKyWunR48e6tGjR2B/4MCBqqmp0V//+td2Fz7OcDI+p0+flsvl0uLFiwO/9Dlr1izdd999evXVVxUbGxv2em1r7Xvzm2++qcsvv1wjRowIU2WR52Rsdu3apd///vf605/+pF/84heqra3VU089pQkTJmj+/Pk2yg27qAof3bt3V4cOHVqkxYaGhhap8oeWLFmihx56SP/4xz80dOjQcJYZMW0Zn/T0dEnSDTfcoPr6ehUWFra78OF0fJqbm7VlyxZt27ZNjz32mKTv/qAYYxQTE6M1a9ZoyJAhVmoPt7a8dr5vwIABevvtt0NdXsS1Zny8Xq+uuuqqoJ8Yv/7662WM0f79+5WRkRHWmm1qy+vHGKM33nhDY8eOVceOHcNZZkS0ZmyKi4t1yy236KmnnpIk9enTR126dNFtt92mZ599Vl6vN+x1h1tU3fPRsWNH9evXT+Xl5UHt5eXlys7OPufj3nnnHY0fP16lpaXt6jOzH2rt+PyQMUZ+vz/U5UWc0/HxeDzasWOHtm/fHtgmTJigHj16aPv27crKyrJVetiF6rWzbdu2dvHG+EOtGZ9bbrlFBw4c0JEjRwJte/bs0SWXXKKUlJSw1mtbW14/lZWV+vzzz/XQQw+Fs8SIac3YHDt2TJdcEvznuUOHDpK+e39uFyJym2sbnPnK0vz5882uXbtMfn6+6dKlS+AO6alTp5qxY8cG+peWlpqYmBjz6quvBn2tq7GxMVJPIaycjk9JSYlZtmyZ2bNnj9mzZ4954403jMfjMdOnT4/UUwgrp+PzQ+352y5Ox+all14yS5cuNXv27DGffvqpmTp1qpFk3n333Ug9hbByOj7Nzc0mJSXF3HfffWbnzp2msrLSZGRkmIcffjhSTyGsWvtv68EHHzRZWVm2y7XK6dgsWLDAxMTEmDlz5pgvvvjCbNy40fTv39/cfPPNkXoKIRd14cMYY1599VWTlpZmOnbsaH7+85+bysrKwLFx48aZQYMGBfYHDRpkJLXYxo0bZ79wS5yMzyuvvGJ69eplOnfubDwej7nxxhvNnDlzzKlTpyJQuR1OxueH2nP4MMbZ2LzwwgvmmmuuMZ06dTJXXHGFufXWW83y5csjULU9Tl87n332mRk6dKiJjY01KSkpZvLkyebYsWOWq7bH6fg0Njaa2NhYM2/ePMuV2ud0bF555RXTs2dPExsba7xer3nggQfM/v37LVcdPi5j2sscDgAAiAZRdc8HAACIfoQPAABgFeEDAABYRfgAAABWET4AAIBVhA8AAGAV4QMAAFhF+AAAAFYRPgAAgFWEDwAAYBXhAwAAWEX4AAAAVv0/QrSACg3FgMwAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAigAAAGdCAYAAAA44ojeAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAdPUlEQVR4nO3df3TV9X348VcwEoJLYtGZH8IctikWndbBSsG2MIX0cHrWuZ6eeYr16Kbn6NCVjLNSGDsz7OyQSTtqHeqmR62nFe1ZV1fPWe2IxxqxzA0UzizZZK3MxgOBYiEJSMPA9/cPD/mahq7c5N7kTfJ4nHMP537uJ/fzep8L5Hk+ufeTspRSCgCAjEwY7QEAAH6eQAEAsiNQAIDsCBQAIDsCBQDIjkABALIjUACA7AgUACA75aM9wM97++23Y8+ePVFVVRVlZWWjPQ4AcBpSStHb2xsNDQ0xYcLwz39kFyh79uyJadOmjfYYAMAQdHZ2xtSpU4f9PNkFSlVVVUS8s8Dq6upRngYAOB09PT0xbdq0/u/jw5VdoJz8sU51dbVAAYAzTLHenuFNsgBAdgQKAJAdgQIAZEegAADZESgAQHYECgCQHYECAGRHoAAA2REoAEB2BAoAkB2BAgBkR6AAANkRKABAdgQKAJCd8tEeAKAQz3TsG5XjLpxZOyrHhfHKGRQAIDsCBQDIjkABALIjUACA7AgUACA7AgUAyI5AAQCyI1AAgOwIFAAgOwIFAMiOQAEAsiNQAIDsCBQAIDsCBQDIjkABALIjUACA7AgUACA7AgUAyI5AAQCyI1AAgOwIFAAgOwIFAMiOQAEAsiNQAIDsCBQAIDsCBQDIjkABALIjUACA7AgUACA7AgUAyI5AAQCyI1AAgOwIFAAgOwIFAMiOQAEAsiNQAIDsCBQAIDsCBQDIjkABALIjUACA7AgUACA7AgUAyI5AAQCyI1AAgOwIFAAgOwIFAMiOQAEAsiNQAIDsDCtQWltbo6ysLJqbm/u3pZSipaUlGhoaorKyMhYsWBA7d+4c7pwAwDgy5EDZunVrPPDAA3H55ZcP2L5u3bpYv359bNiwIbZu3Rp1dXWxaNGi6O3tHfawAMD4MKRAOXz4cFx//fXx4IMPxnve857+7SmluPvuu2P16tXxqU99Ki677LJ49NFH46233oqNGzcWbWgAYGwbUqDcfvvt8YlPfCIWLlw4YPvu3bujq6srmpqa+rdVVFTE/PnzY8uWLad8rr6+vujp6RlwAwDGt/JCv+CJJ56Il19+ObZu3Trosa6uroiIqK2tHbC9trY2Xn/99VM+X2tra6xZs6bQMQCAMaygMyidnZ2xbNmy+PrXvx6TJk36hfuVlZUNuJ9SGrTtpFWrVkV3d3f/rbOzs5CRAIAxqKAzKC+99FLs378/Zs2a1b/txIkT8fzzz8eGDRvi1VdfjYh3zqTU19f377N///5BZ1VOqqioiIqKiqHMDgCMUQWdQbnmmmvilVdeiR07dvTfZs+eHddff33s2LEjLr744qirq4u2trb+rzl27Fi0t7fHvHnzij48ADA2FXQGpaqqKi677LIB284555w477zz+rc3NzfH2rVro7GxMRobG2Pt2rUxefLkWLJkSfGmBgDGtILfJPvLrFixIo4ePRpLly6NgwcPxpw5c2LTpk1RVVVV7EMBAGNUWUopjfYQ79bT0xM1NTXR3d0d1dXVoz0OkJlnOvaNynEXzjz1++iAdxT7+3fRz6AAI8c3a2Cs8ssCAYDsCBQAIDsCBQDIjkABALIjUACA7AgUACA7AgUAyI5AAQCyI1AAgOwIFAAgOwIFAMiOQAEAsiNQAIDsCBQAIDsCBQDIjkABALIjUACA7AgUACA7AgUAyI5AAQCyI1AAgOwIFAAgO+WjPQBw5nmmY99ojwCMcc6gAADZESgAQHYECgCQHYECAGRHoAAA2REoAEB2fMwYIGOj9ZHuhTNrR+W4cJIzKABAdgQKAJAdgQIAZEegAADZESgAQHYECgCQHYECAGTHdVAATsNoXY8ExitnUACA7AgUACA7AgUAyI5AAQCyI1AAgOwIFAAgOwIFAMiOQAEAsiNQAIDsCBQAIDsCBQDIjkABALIjUACA7AgUACA7AgUAyI5AAQCyI1AAgOwIFAAgOwIFAMiOQAEAsiNQAIDsCBQAIDsCBQDIjkABALJTUKDcf//9cfnll0d1dXVUV1fH3Llz4+mnn+5/PKUULS0t0dDQEJWVlbFgwYLYuXNn0YcGAMa2ggJl6tSp8dd//dexbdu22LZtW1x99dXxu7/7u/0Rsm7duli/fn1s2LAhtm7dGnV1dbFo0aLo7e0tyfAAwNhUllJKw3mCKVOmxBe/+MX4wz/8w2hoaIjm5ub4whe+EBERfX19UVtbG3fddVfceuutp/V8PT09UVNTE93d3VFdXT2c0WDMe6Zj32iPwBi1cGbtaI/AGabY37+H/B6UEydOxBNPPBFHjhyJuXPnxu7du6Orqyuampr696moqIj58+fHli1bhj0oADB+lBf6Ba+88krMnTs3fvazn8Wv/MqvxJNPPhkzZ87sj5Da2oHVXVtbG6+//vovfL6+vr7o6+vrv9/T01PoSADAGFPwGZQZM2bEjh074sUXX4w/+qM/ihtvvDE6Ojr6Hy8rKxuwf0pp0LZ3a21tjZqamv7btGnTCh0JABhjCg6UiRMnxvve976YPXt2tLa2xhVXXBFf+cpXoq6uLiIiurq6Buy/f//+QWdV3m3VqlXR3d3df+vs7Cx0JABgjBn2dVBSStHX1xfTp0+Purq6aGtr63/s2LFj0d7eHvPmzfuFX19RUdH/seWTNwBgfCvoPSh/9md/FosXL45p06ZFb29vPPHEE/Hcc8/Fd7/73SgrK4vm5uZYu3ZtNDY2RmNjY6xduzYmT54cS5YsKdX8AMAYVFCg7Nu3L2644YbYu3dv1NTUxOWXXx7f/e53Y9GiRRERsWLFijh69GgsXbo0Dh48GHPmzIlNmzZFVVVVSYYHAMamYV8HpdhcBwVOn+ugUCqug0KhsrkOCgBAqQgUACA7AgUAyI5AAQCyI1AAgOwIFAAgOwIFAMiOQAEAsiNQAIDsCBQAIDsCBQDIjkABALIjUACA7AgUACA7AgUAyI5AAQCyUz7aAwCQn2c69o3KcRfOrB2V45IfZ1AAgOwIFAAgOwIFAMiOQAEAsiNQAIDsCBQAIDsCBQDIjkABALIjUACA7AgUACA7AgUAyI5AAQCyI1AAgOwIFAAgOwIFAMiOQAEAsiNQAIDsCBQAIDsCBQDIjkABALIjUACA7AgUACA75aM9AIwFz3TsG+0RAMYUZ1AAgOwIFAAgOwIFAMiOQAEAsiNQAIDsCBQAIDsCBQDIjkABALIjUACA7AgUACA7AgUAyI5AAQCyI1AAgOwIFAAgOwIFAMiOQAEAsiNQAIDsCBQAIDsCBQDIjkABALIjUACA7AgUACA7AgUAyI5AAQCyU1CgtLa2xm/91m9FVVVVXHDBBXHttdfGq6++OmCflFK0tLREQ0NDVFZWxoIFC2Lnzp1FHRoAGNsKCpT29va4/fbb48UXX4y2trY4fvx4NDU1xZEjR/r3WbduXaxfvz42bNgQW7dujbq6uli0aFH09vYWfXgAYGwqSymloX7xT37yk7jggguivb09Pvaxj0VKKRoaGqK5uTm+8IUvREREX19f1NbWxl133RW33nrrL33Onp6eqKmpie7u7qiurh7qaDCinunYN9ojwJiwcGbtaI/AEBX7+/ew3oPS3d0dERFTpkyJiIjdu3dHV1dXNDU19e9TUVER8+fPjy1btpzyOfr6+qKnp2fADQAY34YcKCmlWL58eXzkIx+Jyy67LCIiurq6IiKitnZgAdfW1vY/9vNaW1ujpqam/zZt2rShjgQAjBFDDpQ77rgj/uM//iMef/zxQY+VlZUNuJ9SGrTtpFWrVkV3d3f/rbOzc6gjAQBjRPlQvuiP//iP46mnnornn38+pk6d2r+9rq4uIt45k1JfX9+/ff/+/YPOqpxUUVERFRUVQxkDABijCjqDklKKO+64I771rW/Fs88+G9OnTx/w+PTp06Ouri7a2tr6tx07diza29tj3rx5xZkYABjzCjqDcvvtt8fGjRvj29/+dlRVVfW/r6SmpiYqKyujrKwsmpubY+3atdHY2BiNjY2xdu3amDx5cixZsqQkCwAAxp6CAuX++++PiIgFCxYM2P7II4/ETTfdFBERK1asiKNHj8bSpUvj4MGDMWfOnNi0aVNUVVUVZWAAYOwb1nVQSsF1UDgTuQ4KFIfroJy5sroOCgBAKQgUACA7AgUAyI5AAQCyI1AAgOwIFAAgOwIFAMiOQAEAsjOkXxYIuXLBNICxwRkUACA7AgUAyI5AAQCyI1AAgOwIFAAgOwIFAMiOQAEAsiNQAIDsCBQAIDsCBQDIjkABALIjUACA7AgUACA7AgUAyE75aA/A2PRMx77RHgGAM5gzKABAdgQKAJAdgQIAZEegAADZESgAQHYECgCQHYECAGTHdVAASuT8Pc+O2LEONFw9YseCkeAMCgCQHYECAGRHoAAA2REoAEB2BAoAkB2BAgBkR6AAANkRKABAdgQKAJAdgQIAZEegAADZESgAQHYECgCQHYECAGSnfLQHABgp5+95drRH4Jd4pmPfqBx34czaUTkuv5gzKABAdgQKAJAdgQIAZEegAADZESgAQHYECgCQHYECAGRHoAAA2REoAEB2BAoAkB2BAgBkR6AAANkRKABAdgQKAJAdgQIAZEegAADZESgAQHYKDpTnn38+fud3ficaGhqirKws/umf/mnA4ymlaGlpiYaGhqisrIwFCxbEzp07izUvADAOFBwoR44ciSuuuCI2bNhwysfXrVsX69evjw0bNsTWrVujrq4uFi1aFL29vcMeFgAYH8oL/YLFixfH4sWLT/lYSinuvvvuWL16dXzqU5+KiIhHH300amtrY+PGjXHrrbcOb1oAYFwo6ntQdu/eHV1dXdHU1NS/raKiIubPnx9btmw55df09fVFT0/PgBsAML4VfAbl/9LV1RUREbW1tQO219bWxuuvv37Kr2ltbY01a9YUcwze5ZmOfSN6vPP3PPvOnyN0vAMNV4/QkQAYSSX5FE9ZWdmA+ymlQdtOWrVqVXR3d/ffOjs7SzESAHAGKeoZlLq6uoh450xKfX19//b9+/cPOqtyUkVFRVRUVBRzDADgDFfUMyjTp0+Purq6aGtr69927NixaG9vj3nz5hXzUADAGFbwGZTDhw/HD3/4w/77u3fvjh07dsSUKVPi137t16K5uTnWrl0bjY2N0djYGGvXro3JkyfHkiVLijo4ADB2FRwo27Zti9/+7d/uv798+fKIiLjxxhvjq1/9aqxYsSKOHj0aS5cujYMHD8acOXNi06ZNUVVVVbypAYAxreBAWbBgQaSUfuHjZWVl0dLSEi0tLcOZCwAYx/wuHgAgO0X9FA9QXCevKzNWuY5N8Yz03xWvHaXmDAoAkB2BAgBkR6AAANkRKABAdgQKAJAdgQIAZMfHjIFRM9Y/Rg0MnTMoAEB2BAoAkB2BAgBkR6AAANkRKABAdgQKAJAdgQIAZEegAADZESgAQHYECgCQHYECAGRHoAAA2REoAEB2BAoAkJ3y0R4AhuP8Pc+O6PEONFw9oscDGK+cQQEAsiNQAIDsCBQAIDsCBQDIjkABALIjUACA7AgUACA7roMCQPZKfs2js84deH/G4tIej1/KGRQAIDsCBQDIjkABALIjUACA7AgUACA7AgUAyI6PGUMBSv5RR2BceaZj36gcd+HM2lE5biGcQQEAsiNQAIDsCBQAIDsCBQDIjkABALIjUACA7AgUACA7roMCQMFcE4hScwYFAMiOQAEAsiNQAIDsCBQAIDsCBQDIjkABALLjY8YAjHs7Og8N3ND5+Igc9/wROcopzPzMaB35tDmDAgBkR6AAANkRKABAdgQKAJAdgQIAZEegAADZESgAQHbG3XVQnunYN9ojjOivKR+1z9gDwDA4gwIAZEegAADZKVmg3HfffTF9+vSYNGlSzJo1KzZv3lyqQwEAY0xJAuUb3/hGNDc3x+rVq2P79u3x0Y9+NBYvXhw//vGPS3E4AGCMKUmgrF+/Pm6++ea45ZZb4gMf+EDcfffdMW3atLj//vtLcTgAYIwp+qd4jh07Fi+99FKsXLlywPampqbYsmXLoP37+vqir6+v/353d3dERPT09BR7tIiIOHK4tyTPW4hJR94a7REAGMdK8T325HOmlIryfEUPlAMHDsSJEyeitrZ2wPba2tro6uoatH9ra2usWbNm0PZp06YVezQAICIibinZM/f29kZNTc2wn6dk10EpKysbcD+lNGhbRMSqVati+fLl/ffffvvt+OlPfxrnnXde//49PT0xbdq06OzsjOrq6lKNnJXxuOaI8blua7bmsWo8rjlifK775Jo7OjqioaGhKM9Z9EA5//zz46yzzhp0tmT//v2DzqpERFRUVERFRcWAbeeee+4pn7u6unrcvNgnjcc1R4zPdVvz+GDN48d4XPeFF14YEyYU5+2tRX+T7MSJE2PWrFnR1tY2YHtbW1vMmzev2IcDAMagkvyIZ/ny5XHDDTfE7NmzY+7cufHAAw/Ej3/847jttttKcTgAYIwpSaBcd9118eabb8Zf/uVfxt69e+Oyyy6L73znO3HRRRcN6fkqKirizjvvHPSjoLFsPK45Ynyu25rHB2seP8bjukux5rJUrM8DAQAUid/FAwBkR6AAANkRKABAdgQKAJCdbALlvvvui+nTp8ekSZNi1qxZsXnz5v9z//b29pg1a1ZMmjQpLr744vi7v/u7EZq0eApZ8969e2PJkiUxY8aMmDBhQjQ3N4/coEVUyJq/9a1vxaJFi+JXf/VXo7q6OubOnRv/8i//MoLTFk8h637hhRfiqquuivPOOy8qKyvjkksuiS9/+csjOG1xFPpv+qTvf//7UV5eHh/84AdLO2AJFLLm5557LsrKygbd/uu//msEJx6+Ql/nvr6+WL16dVx00UVRUVER733ve+Phhx8eoWmLo5A133TTTad8nS+99NIRnLg4Cn2tH3vssbjiiiti8uTJUV9fH3/wB38Qb7755ukfMGXgiSeeSGeffXZ68MEHU0dHR1q2bFk655xz0uuvv37K/V977bU0efLktGzZstTR0ZEefPDBdPbZZ6dvfvObIzz50BW65t27d6fPfe5z6dFHH00f/OAH07Jly0Z24CIodM3Lli1Ld911V/r3f//3tGvXrrRq1ap09tlnp5dffnmEJx+eQtf98ssvp40bN6Yf/OAHaffu3elrX/tamjx5cvr7v//7EZ586Apd80mHDh1KF198cWpqakpXXHHFyAxbJIWu+Xvf+16KiPTqq6+mvXv39t+OHz8+wpMP3VBe509+8pNpzpw5qa2tLe3evTv927/9W/r+978/glMPT6FrPnTo0IDXt7OzM02ZMiXdeeedIzv4MBW67s2bN6cJEyakr3zlK+m1115LmzdvTpdeemm69tprT/uYWQTKhz70oXTbbbcN2HbJJZeklStXnnL/FStWpEsuuWTAtltvvTV9+MMfLtmMxVbomt9t/vz5Z2SgDGfNJ82cOTOtWbOm2KOVVDHW/Xu/93vps5/9bLFHK5mhrvm6665Lf/7nf57uvPPOMy5QCl3zyUA5ePDgCExXGoWu+emnn041NTXpzTffHInxSmK4/56ffPLJVFZWlv7nf/6nFOOVTKHr/uIXv5guvvjiAdvuueeeNHXq1NM+5qj/iOfYsWPx0ksvRVNT04DtTU1NsWXLllN+zb/+678O2v/jH/94bNu2Lf73f/+3ZLMWy1DWfKYrxprffvvt6O3tjSlTppRixJIoxrq3b98eW7Zsifnz55dixKIb6pofeeSR+NGPfhR33nlnqUcsuuG8zldeeWXU19fHNddcE9/73vdKOWZRDWXNTz31VMyePTvWrVsXF154Ybz//e+PP/3TP42jR4+OxMjDVox/zw899FAsXLhwyBcuHQ1DWfe8efPijTfeiO985zuRUop9+/bFN7/5zfjEJz5x2sct2W8zPl0HDhyIEydODPpFgrW1tYN+4eBJXV1dp9z/+PHjceDAgaivry/ZvMUwlDWf6Yqx5r/5m7+JI0eOxO///u+XYsSSGM66p06dGj/5yU/i+PHj0dLSErfcUrpfj15MQ1nzf//3f8fKlStj8+bNUV4+6v8tFWwoa66vr48HHnggZs2aFX19ffG1r30trrnmmnjuuefiYx/72EiMPSxDWfNrr70WL7zwQkyaNCmefPLJOHDgQCxdujR++tOfnhHvQxnu/2N79+6Np59+OjZu3FiqEUtiKOueN29ePPbYY3HdddfFz372szh+/Hh88pOfjL/927897eNm8z9BWVnZgPsppUHbftn+p9qes0LXPBYMdc2PP/54tLS0xLe//e244IILSjVeyQxl3Zs3b47Dhw/Hiy++GCtXroz3ve998ZnPfKaUYxbV6a75xIkTsWTJklizZk28//3vH6nxSqKQ13nGjBkxY8aM/vtz586Nzs7O+NKXvnRGBMpJhaz57bffjrKysnjssceipqYmIiLWr18fn/70p+Pee++NysrKks9bDEP9f+yrX/1qnHvuuXHttdeWaLLSKmTdHR0d8bnPfS7+4i/+Ij7+8Y/H3r174/Of/3zcdttt8dBDD53W8UY9UM4///w466yzBlXY/v37B9XaSXV1dafcv7y8PM4777ySzVosQ1nzmW44a/7GN74RN998c/zDP/xDLFy4sJRjFt1w1j19+vSIiPiN3/iN2LdvX7S0tJwRgVLomnt7e2Pbtm2xffv2uOOOOyLinW9kKaUoLy+PTZs2xdVXXz0isw9Vsf5Nf/jDH46vf/3rxR6vJIay5vr6+rjwwgv74yQi4gMf+ECklOKNN96IxsbGks48XMN5nVNK8fDDD8cNN9wQEydOLOWYRTeUdbe2tsZVV10Vn//85yMi4vLLL49zzjknPvrRj8Zf/dVfndZPOkb9PSgTJ06MWbNmRVtb24DtbW1tMW/evFN+zdy5cwftv2nTppg9e3acffbZJZu1WIay5jPdUNf8+OOPx0033RQbN24s6GeXuSjWa51Sir6+vmKPVxKFrrm6ujpeeeWV2LFjR//ttttuixkzZsSOHTtizpw5IzX6kBXrdd6+fXv2P6I+aShrvuqqq2LPnj1x+PDh/m27du2KCRMmxNSpU0s6bzEM53Vub2+PH/7wh3HzzTeXcsSSGMq633rrrZgwYWBinHXWWRHx/3/i8Uud9ttpS+jkx5ceeuih1NHRkZqbm9M555zT/y7nlStXphtuuKF//5MfM/6TP/mT1NHRkR566KEz9mPGp7vmlFLavn172r59e5o1a1ZasmRJ2r59e9q5c+dojD8kha5548aNqby8PN17770DPqZ36NCh0VrCkBS67g0bNqSnnnoq7dq1K+3atSs9/PDDqbq6Oq1evXq0llCwofz9frcz8VM8ha75y1/+cnryySfTrl270g9+8IO0cuXKFBHpH//xH0drCQUrdM29vb1p6tSp6dOf/nTauXNnam9vT42NjemWW24ZrSUUbKh/tz/72c+mOXPmjPS4RVPouh955JFUXl6e7rvvvvSjH/0ovfDCC2n27NnpQx/60GkfM4tASSmle++9N1100UVp4sSJ6Td/8zdTe3t7/2M33nhjmj9//oD9n3vuuXTllVemiRMnpl//9V9P999//whPPHyFrjkiBt0uuuiikR16mApZ8/z580+55htvvHHkBx+mQtZ9zz33pEsvvTRNnjw5VVdXpyuvvDLdd9996cSJE6Mw+dAV+vf73c7EQEmpsDXfdddd6b3vfW+aNGlSes973pM+8pGPpH/+538ehamHp9DX+T//8z/TwoULU2VlZZo6dWpavnx5euutt0Z46uEpdM2HDh1KlZWV6YEHHhjhSYur0HXfc889aebMmamysjLV19en66+/Pr3xxhunfbyylE73XAsAwMgY9fegAAD8PIECAGRHoAAA2REoAEB2BAoAkB2BAgBkR6AAANkRKABAdgQKAJAdgQIAZEegAADZESgAQHb+H2xKirMNCXhYAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mem_with_na = df['membership']\n",
    "mem_without_na = not_null_df['membership']\n",
    "trait_with_na = df['traitsig']\n",
    "trait_without_na= not_null_df['traitsig']\n",
    "\n",
    "\n",
    "\n",
    "plt.hist(x=df['membership'], bins=15, alpha= .3)\n",
    "plt.hist(x=not_null_df['membership'], bins=15, alpha = .3)\n",
    "plt.show()\n",
    "\n",
    "plt.hist(x=df['traitsig'], bins=15, alpha= .3)\n",
    "plt.hist(x=not_null_df['traitsig'], bins=15, alpha = .3)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f46facf3",
   "metadata": {},
   "source": [
    "2.) Using NCBI datasets, create an enzyme classification on available genes. Note, not every gene has an enzyme directly produced.  Second note, genes that were not in this dataset were dropped, and genes without an enzyme produced in the description were dropped."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b3b656c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.11/site-packages/Bio/Application/__init__.py:40: BiopythonDeprecationWarning: The Bio.Application modules and modules relying on it have been deprecated.\n",
      "\n",
      "Due to the on going maintenance burden of keeping command line application\n",
      "wrappers up to date, we have decided to deprecate and eventually remove these\n",
      "modules.\n",
      "\n",
      "We instead now recommend building your command line and invoking it directly\n",
      "with the subprocess module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "from collections import defaultdict\n",
    "from io import BytesIO, StringIO\n",
    "from pathlib import Path\n",
    "from zipfile import ZipFile\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio import AlignIO, Phylo, SeqIO\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "from Bio.Phylo.TreeConstruction import (\n",
    "    NNITreeSearcher,\n",
    "    ParsimonyScorer,\n",
    "    ParsimonyTreeConstructor,\n",
    ")\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from ncbi.datasets import GeneApi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "bc7ffa6a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'list'>\n"
     ]
    }
   ],
   "source": [
    "taxon = 'human'\n",
    "symbols = df['gene'].tolist()\n",
    "print (type(symbols))\n",
    "\n",
    "gene_api = GeneApi()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d6c0db39",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = gene_api.gene_metadata_by_tax_and_symbol(symbols, taxon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2b9d51a8",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute 'description'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[15], line 5\u001b[0m\n\u001b[1;32m      1\u001b[0m gene_des \u001b[38;5;241m=\u001b[39m {}\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m gene_data \u001b[38;5;129;01min\u001b[39;00m metadata\u001b[38;5;241m.\u001b[39mgenes:\n\u001b[0;32m----> 5\u001b[0m     x \u001b[38;5;241m=\u001b[39m gene_data\u001b[38;5;241m.\u001b[39mgene\u001b[38;5;241m.\u001b[39mdescription\n\u001b[1;32m      6\u001b[0m     y \u001b[38;5;241m=\u001b[39m gene_data\u001b[38;5;241m.\u001b[39mgene\u001b[38;5;241m.\u001b[39msymbol\n\u001b[1;32m      8\u001b[0m     gene_des[y] \u001b[38;5;241m=\u001b[39m x\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'NoneType' object has no attribute 'description'"
     ]
    }
   ],
   "source": [
    "gene_des = {}\n",
    "\n",
    "for gene_data in metadata.genes:\n",
    "    \n",
    "    x = gene_data.gene.description\n",
    "    y = gene_data.gene.symbol\n",
    "    \n",
    "    gene_des[y] = x\n",
    "    \n",
    "### an attribute error arises but it is fine please proceed    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "554985b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#print (gene_des)\n",
    "#print (len(gene_des))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0aa63cf0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        gene                                  enzyme_des\n",
      "0     HOTAIR                HOX transcript antisense RNA\n",
      "1  LINC00892  long intergenic non-protein coding RNA 892\n",
      "2  LINC00605  long intergenic non-protein coding RNA 605\n",
      "3     IPO5P1                     importin 5 pseudogene 1\n",
      "4   GOLGA8DP       golgin A8 family member D, pseudogene\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>enzyme_des</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>264</td>\n",
       "      <td>264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>264</td>\n",
       "      <td>264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>HOTAIR</td>\n",
       "      <td>HOX transcript antisense RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene                    enzyme_des\n",
       "count      264                           264\n",
       "unique     264                           264\n",
       "top     HOTAIR  HOX transcript antisense RNA\n",
       "freq         1                             1"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "keys = []\n",
    "values = []\n",
    "\n",
    "for key,value in gene_des.items():\n",
    "    keys.append(key)\n",
    "    values.append(value)\n",
    "    \n",
    "\n",
    "data = {'gene': keys,\n",
    "        'enzyme_des': values}\n",
    "\n",
    "df_des = pd.DataFrame(data)\n",
    "\n",
    "\n",
    "print (df_des.head())\n",
    "df_des.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14976d0c",
   "metadata": {},
   "source": [
    "The majority of enzymes are identified with the suffix 'ase', so I used this string to search for candidates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e66aae13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'deaminase', 'other', 'other', 'carboxypeptidase', 'other', 'kinase', 'other', 'kinase', 'other', 'metallopeptidase', 'other', 'other', 'other', 'other', 'other', 'kinase', 'anhydrase', 'cyclase', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'kinase', 'other', 'other', 'other', 'other', 'other', 'kinase', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'dehydrogenase', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'kinase', 'other', 'GTPase', 'other', 'alpha-L-fucosidase', 'kinase', 'other', 'beta-secretase', 'other', 'other', 'other', 'other', 'other', 'other', 'glutaminase', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'peptidase', 'other', 'other', 'other', 'desaturase', 'other', 'other', 'oxidase', 'other', 'other', 'other', 'other', 'other', '6-beta-N-acetylglucosaminyltransferase', 'other', 'other', 'sulfotransferase', 'other', 'other', 'other', 'peroxidase', 'synthetase', 'oxidase', 'other', 'other', 'other', 'other', 'other', 'other', 'phosphodiesterase', 'other', 'other', 'other', 'other', 'scramblase', 'other', 'other', 'other', 'lyase', 'other', 'other', 'isopentenyltransferase', 'other', 'synthase', 'phosphatase', 'GTPase,', 'kinase', 'other', 'helicase', 'other', 'kinase', 'synthetase', 'other', 'dihydropyrimidinase', 'other', 'other', 'ATPase', 'other', 'other', 'other', 'other', 'phosphatase', 'other', 'endonuclease', 'other', 'other', 'reductase', 'other', 'peptidase', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'peptidase', 'other', 'other', 'other', 'other', 'other', 'glucosyltransferase', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'caspase', 'other', 'other', 'phosphatase', 'other', 'other', 'synthetase', 'other', 'other', 'other', 'other', 'other', '4-hydroxylase', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'other', 'dehydrogenase/reductase', 'other', 'other', 'other', 'peptidase', 'other', 'other', 'other', 'other', 'GTPase', 'GTPase', 'other']\n",
      "264\n"
     ]
    }
   ],
   "source": [
    "enz_list = []\n",
    "\n",
    "for x in df_des['enzyme_des']:\n",
    "    ase = 'ase'\n",
    "    li_st = [i for i in x.split() if ase in i ]\n",
    "    if li_st == []:\n",
    "        li_st_ = 'other'\n",
    "    else:\n",
    "        li_st_ = li_st[0]\n",
    "    \n",
    "    enz_list.append(li_st_)\n",
    "        \n",
    "print (enz_list)\n",
    "print(len(enz_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "cff21fb4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        gene assoc_enzymes\n",
      "0     HOTAIR         other\n",
      "1  LINC00892         other\n",
      "2  LINC00605         other\n",
      "3     IPO5P1         other\n",
      "4   GOLGA8DP         other\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>assoc_enzymes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>264</td>\n",
       "      <td>264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>264</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>HOTAIR</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1</td>\n",
       "      <td>212</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene assoc_enzymes\n",
       "count      264           264\n",
       "unique     264            35\n",
       "top     HOTAIR         other\n",
       "freq         1           212"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_2 = {'gene': keys,\n",
    "        'assoc_enzymes': enz_list}\n",
    "\n",
    "df_enz = pd.DataFrame(data_2)\n",
    "\n",
    "\n",
    "print (df_enz.head())\n",
    "df_enz.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "75fd0fcc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>assoc_enzymes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>52</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>52</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>ADAR</td>\n",
       "      <td>kinase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene assoc_enzymes\n",
       "count     52            52\n",
       "unique    52            34\n",
       "top     ADAR        kinase\n",
       "freq       1             9"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_enz_without_other = df_enz[df_enz['assoc_enzymes'] != 'other']\n",
    "\n",
    "df_enz_without_other.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "c5d1519e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>membership</th>\n",
       "      <th>traitsig</th>\n",
       "      <th>ensg_version</th>\n",
       "      <th>gene</th>\n",
       "      <th>LV_number</th>\n",
       "      <th>LV_pathway</th>\n",
       "      <th>assoc_enzymes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.5158</td>\n",
       "      <td>0.2172</td>\n",
       "      <td>ENSG00000137959.15</td>\n",
       "      <td>IFI44L</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.5497</td>\n",
       "      <td>0.4713</td>\n",
       "      <td>ENSG00000160710.16</td>\n",
       "      <td>ADAR</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "      <td>deaminase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.7754</td>\n",
       "      <td>0.5447</td>\n",
       "      <td>ENSG00000082512.14</td>\n",
       "      <td>TRAF5</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.7309</td>\n",
       "      <td>0.4325</td>\n",
       "      <td>ENSG00000163219.11</td>\n",
       "      <td>ARHGAP25</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "      <td>GTPase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.5587</td>\n",
       "      <td>0.4396</td>\n",
       "      <td>ENSG00000153208.16</td>\n",
       "      <td>MERTK</td>\n",
       "      <td>2</td>\n",
       "      <td>MetaMEx_AerobicTraining</td>\n",
       "      <td>kinase</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   membership  traitsig        ensg_version      gene LV_number  \\\n",
       "0      0.5158    0.2172  ENSG00000137959.15    IFI44L         2   \n",
       "1      0.5497    0.4713  ENSG00000160710.16      ADAR         2   \n",
       "2      0.7754    0.5447  ENSG00000082512.14     TRAF5         2   \n",
       "3      0.7309    0.4325  ENSG00000163219.11  ARHGAP25         2   \n",
       "4      0.5587    0.4396  ENSG00000153208.16     MERTK         2   \n",
       "\n",
       "                LV_pathway assoc_enzymes  \n",
       "0  MetaMEx_AerobicTraining         other  \n",
       "1  MetaMEx_AerobicTraining     deaminase  \n",
       "2  MetaMEx_AerobicTraining         other  \n",
       "3  MetaMEx_AerobicTraining        GTPase  \n",
       "4  MetaMEx_AerobicTraining        kinase  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_2 = df.merge(df_enz, how='inner', on='gene')\n",
    "\n",
    "df_2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1fc849f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>membership</th>\n",
       "      <th>traitsig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>252.000000</td>\n",
       "      <td>252.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.584875</td>\n",
       "      <td>0.433260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.114327</td>\n",
       "      <td>0.137893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.236800</td>\n",
       "      <td>0.029100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.493000</td>\n",
       "      <td>0.345375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>0.589450</td>\n",
       "      <td>0.448900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.678075</td>\n",
       "      <td>0.532750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>0.838100</td>\n",
       "      <td>0.765800</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       membership    traitsig\n",
       "count  252.000000  252.000000\n",
       "mean     0.584875    0.433260\n",
       "std      0.114327    0.137893\n",
       "min      0.236800    0.029100\n",
       "25%      0.493000    0.345375\n",
       "50%      0.589450    0.448900\n",
       "75%      0.678075    0.532750\n",
       "max      0.838100    0.765800"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_2.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8772e0e",
   "metadata": {},
   "source": [
    "There were two genes, containing enzymes function, that were lost in the join, so I had to see where the NCBI dataset changed the gene name, and then manually insert the enzymes into those genes who had their name changed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "91fc6a2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>membership</th>\n",
       "      <th>traitsig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>50.000000</td>\n",
       "      <td>50.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.586992</td>\n",
       "      <td>0.406314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.110572</td>\n",
       "      <td>0.143765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.303700</td>\n",
       "      <td>0.082200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.497425</td>\n",
       "      <td>0.313175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>0.602200</td>\n",
       "      <td>0.396450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.677300</td>\n",
       "      <td>0.513150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>0.760100</td>\n",
       "      <td>0.639500</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       membership   traitsig\n",
       "count   50.000000  50.000000\n",
       "mean     0.586992   0.406314\n",
       "std      0.110572   0.143765\n",
       "min      0.303700   0.082200\n",
       "25%      0.497425   0.313175\n",
       "50%      0.602200   0.396450\n",
       "75%      0.677300   0.513150\n",
       "max      0.760100   0.639500"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_2_without_other = df_2[df_2['assoc_enzymes'] != 'other']\n",
    "\n",
    "df_2_without_other.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "1c236955",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2        IFI44L\n",
      "3          ADAR\n",
      "4         TRAF5\n",
      "5      ARHGAP25\n",
      "6         MERTK\n",
      "         ...   \n",
      "259      RASSF2\n",
      "260      TRANK1\n",
      "261    RABGAP1L\n",
      "262    ARHGAP25\n",
      "263       PTBP3\n",
      "Name: gene, Length: 571, dtype: object\n"
     ]
    }
   ],
   "source": [
    "s1 = df['gene']\n",
    "s2 = df_enz['gene']\n",
    "df_compare = pd.concat([s1,s2])\n",
    "\n",
    "print(df_compare)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "068e7c55",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "252\n"
     ]
    }
   ],
   "source": [
    "print(len(df_compare[df_compare.duplicated() == True]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "51c79c42",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ACP3', 'TLCD5', 'DYNLT4', 'TASL', 'IFTAP', 'H2BC26', 'FCGR1BP', 'CFAP91', 'DIPK1C', 'SLC66A3', 'SEPTIN6', 'PABIR2'}\n"
     ]
    }
   ],
   "source": [
    "duplicated = df_compare[df_compare.duplicated() == True].tolist()\n",
    "duplicated_set = set(duplicated)\n",
    "\n",
    "ncbi_gene_list = df_enz['gene'].tolist()\n",
    "ncbi_gene_set = set(ncbi_gene_list)\n",
    "\n",
    "diff = ncbi_gene_set - duplicated_set\n",
    "\n",
    "print (diff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "08e041af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'AC106795.5', 'FAM122B', 'AL365209.1', 'AC244502.1', 'AP001646.2', 'AF117829.1', 'AC011511.4', 'AC011477.1', 'AC004847.1', 'TMEM136', 'AF127936.2', 'AL021328.1', 'AC114271.1', 'AC131097.4', 'AC091488.1', 'CXorf21', 'AP000525.1', 'AC131902.2', 'SEPT6', 'AC125611.3', 'AL353746.1', 'TCTEX1D4', 'FAM69C', 'AC069209.1', 'AC104772.1', 'AC019257.7', 'AC104663.1', 'HIST3H2BB', 'AC098820.2', 'AC113382.2', 'AD000864.1', 'AC245100.4', 'PQLC3', 'AL359762.1', 'AL109811.2', 'AL034397.3', 'AL731533.2', 'AL161785.1', 'AC027702.1', 'MAATS1', 'AL133355.1', 'FP565260.3', 'AL031595.1', 'AC012485.3', 'AL606760.1', 'AC008581.1', 'C11orf74', 'AC105749.1', 'AC020915.3', 'AL445472.1', 'AC012313.2', 'AC099548.2', 'AC098484.3', 'ACPP', 'FCGR1B'}\n"
     ]
    }
   ],
   "source": [
    "orig_gene_list = df['gene'].tolist()\n",
    "orig_gene_set = set(orig_gene_list)\n",
    "\n",
    "ncbi_diff = orig_gene_set - ncbi_gene_set\n",
    "\n",
    "print (ncbi_diff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "c29c517e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        gene assoc_enzymes\n",
      "46     IFTAP         other\n",
      "50    DIPK1C        kinase\n",
      "52    H2BC26         other\n",
      "55   SLC66A3         other\n",
      "67    PABIR2         other\n",
      "82     TLCD5         other\n",
      "85   FCGR1BP         other\n",
      "88   SEPTIN6         other\n",
      "125   DYNLT4         other\n",
      "176     ACP3   phosphatase\n",
      "222     TASL         other\n",
      "242   CFAP91         other\n"
     ]
    }
   ],
   "source": [
    "diff_as_a_list = list(diff)\n",
    "\n",
    "df_enz_without_other_and_set = df_enz[df_enz['gene'].isin(diff_as_a_list)]\n",
    "\n",
    "print (df_enz_without_other_and_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "6f3b726d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index([255], dtype='int64')\n",
      "1\n",
      "membership                 0.4532\n",
      "traitsig                   0.3218\n",
      "ensg_version    ENSG00000187773.8\n",
      "gene                       FAM69C\n",
      "LV_number                     NaN\n",
      "LV_pathway                    NaN\n",
      "Name: 255, dtype: object\n",
      "Index([143], dtype='int64')\n",
      "1\n",
      "membership                  0.6015\n",
      "traitsig                    0.4211\n",
      "ensg_version    ENSG00000014257.15\n",
      "gene                          ACPP\n",
      "LV_number                      NaN\n",
      "LV_pathway                     NaN\n",
      "Name: 143, dtype: object\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/w4/swg_x0mn7sl_p9j90gy4tmk80000gn/T/ipykernel_24469/3250359206.py:15: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  df_2.loc[len(df_2)] = FAM69C_row\n",
      "/var/folders/w4/swg_x0mn7sl_p9j90gy4tmk80000gn/T/ipykernel_24469/3250359206.py:19: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  df_2.loc[len(df_2)] = ACPP_row\n"
     ]
    }
   ],
   "source": [
    "row_number_FAM69C = df[df['gene'] == 'FAM69C'].index\n",
    "\n",
    "print(row_number_FAM69C)\n",
    "\n",
    "print (df.iloc[253])\n",
    "\n",
    "row_number_ACPP = df[df['gene'] == 'ACPP'].index\n",
    "\n",
    "print(row_number_ACPP)\n",
    "\n",
    "print (df.iloc[141])\n",
    "\n",
    "FAM69C_row = {'membership': 0.4532, 'traitsig': 0.3218, 'ensg_version': 'ENSG00000187773.8', 'gene': 'FAM69C', 'LV_number': None, 'LV_pathway':None , 'assoc_enzymes':'kinase'}\n",
    "\n",
    "df_2.loc[len(df_2)] = FAM69C_row\n",
    "\n",
    "ACPP_row = {'membership': 0.6015, 'traitsig': 0.4211, 'ensg_version': 'ENSG00000014257.15', 'gene': 'ACPP', 'LV_number':None , 'LV_pathway':None , 'assoc_enzymes':'phosphatase'}\n",
    "\n",
    "df_2.loc[len(df_2)] = ACPP_row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "b39449e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>membership</th>\n",
       "      <th>traitsig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>52.000000</td>\n",
       "      <td>52.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.584698</td>\n",
       "      <td>0.404973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.109984</td>\n",
       "      <td>0.141422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.303700</td>\n",
       "      <td>0.082200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.495275</td>\n",
       "      <td>0.316325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>0.601500</td>\n",
       "      <td>0.396450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.675900</td>\n",
       "      <td>0.512850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>0.760100</td>\n",
       "      <td>0.639500</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       membership   traitsig\n",
       "count   52.000000  52.000000\n",
       "mean     0.584698   0.404973\n",
       "std      0.109984   0.141422\n",
       "min      0.303700   0.082200\n",
       "25%      0.495275   0.316325\n",
       "50%      0.601500   0.396450\n",
       "75%      0.675900   0.512850\n",
       "max      0.760100   0.639500"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_2_without_other = df_2[df_2['assoc_enzymes'] != 'other']\n",
    "\n",
    "df_2_without_other.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "b2abf0ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     membership  traitsig        ensg_version    gene LV_number LV_pathway  \\\n",
      "249      0.4009    0.4590   ENSG00000183837.9   PNMA3       NaN        NaN   \n",
      "250      0.6261    0.6430  ENSG00000160219.11    GAB3       NaN        NaN   \n",
      "251      0.5075    0.5812   ENSG00000155961.4  RAB39B       NaN        NaN   \n",
      "252      0.4532    0.3218   ENSG00000187773.8  FAM69C       NaN       None   \n",
      "253      0.6015    0.4211  ENSG00000014257.15    ACPP       NaN       None   \n",
      "\n",
      "    assoc_enzymes  \n",
      "249         other  \n",
      "250         other  \n",
      "251         other  \n",
      "252        kinase  \n",
      "253   phosphatase  \n"
     ]
    }
   ],
   "source": [
    "print(df_2.tail())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "14aa43eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('float64')"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_2['membership'].dtype\n",
    "df_2['traitsig'].dtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "2db4ed5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['other' 'deaminase' 'GTPase' 'kinase'\n",
      " '6-beta-N-acetylglucosaminyltransferase' 'scramblase' 'ATPase' 'oxidase'\n",
      " 'GTPase,' 'metallopeptidase' 'caspase' 'synthetase' 'phosphatase'\n",
      " 'beta-secretase' 'dehydrogenase' 'glutaminase' 'cyclase' 'endonuclease'\n",
      " 'desaturase' 'peroxidase' 'glucosyltransferase' '4-hydroxylase'\n",
      " 'isopentenyltransferase' 'reductase' 'dihydropyrimidinase' 'peptidase'\n",
      " 'helicase' 'lyase' 'alpha-L-fucosidase' 'carboxypeptidase'\n",
      " 'sulfotransferase' 'synthase' 'dehydrogenase/reductase'\n",
      " 'phosphodiesterase' 'anhydrase']\n",
      "35\n"
     ]
    }
   ],
   "source": [
    "print(df_2['assoc_enzymes'].unique())\n",
    "print (len(df_2['assoc_enzymes'].unique()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e5505b9",
   "metadata": {},
   "source": [
    "Now I classified the enzymes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "73d45e3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "oxidoreductase = ['oxidase', 'dehydrogenase', 'desaturase', 'peroxidase', '4-hydroxylase', 'reductase', 'dehydrogenase/reductase']\n",
    "transferase = ['kinase','6-beta-N-acetylglucosaminyltransferase','glucosyltransferase', 'isopentenyltransferase', 'sulfotransferase']\n",
    "hydrolase = ['deaminase', 'GTPase', 'GTPase,', 'ATPase', 'metallopeptidase', 'caspase', 'phosphatase', 'beta-secretase', 'glutaminase', 'endonuclease', 'dihydropyrimidinase', 'peptidase', 'helicase', 'alpha-L-fucosidase', 'carboxypeptidase', 'phosphodiesterase']\n",
    "lyase = ['cyclase', 'lyase', 'anhydrase']\n",
    "isomerase = []\n",
    "ligase = ['synthetase', 'synthase']\n",
    "translocase = ['scramblase']\n",
    "\n",
    "ec = [oxidoreductase, transferase, hydrolase, lyase, isomerase, ligase, translocase]\n",
    "\n",
    "enzyme_class = []\n",
    "\n",
    "for x in range(len(df_2['assoc_enzymes'])):\n",
    "    if df_2['assoc_enzymes'][x] in ec[0]:\n",
    "        enzyme_class.append(1)\n",
    "    elif df_2['assoc_enzymes'][x] in ec[1]:\n",
    "        enzyme_class.append(2)\n",
    "    elif df_2['assoc_enzymes'][x] in ec[2]:\n",
    "        enzyme_class.append(3)\n",
    "    elif df_2['assoc_enzymes'][x] in ec[3]:\n",
    "        enzyme_class.append(4)\n",
    "    elif df_2['assoc_enzymes'][x] in ec[5]:\n",
    "        enzyme_class.append(6)\n",
    "    elif df_2['assoc_enzymes'][x] in ec[6]:\n",
    "        enzyme_class.append(7)\n",
    "    else:\n",
    "        enzyme_class.append(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "e8af42fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   membership  traitsig        ensg_version      gene LV_number  \\\n",
      "0      0.5158    0.2172  ENSG00000137959.15    IFI44L         2   \n",
      "1      0.5497    0.4713  ENSG00000160710.16      ADAR         2   \n",
      "2      0.7754    0.5447  ENSG00000082512.14     TRAF5         2   \n",
      "3      0.7309    0.4325  ENSG00000163219.11  ARHGAP25         2   \n",
      "4      0.5587    0.4396  ENSG00000153208.16     MERTK         2   \n",
      "\n",
      "                LV_pathway assoc_enzymes  enzyme_class  \n",
      "0  MetaMEx_AerobicTraining         other             0  \n",
      "1  MetaMEx_AerobicTraining     deaminase             3  \n",
      "2  MetaMEx_AerobicTraining         other             0  \n",
      "3  MetaMEx_AerobicTraining        GTPase             3  \n",
      "4  MetaMEx_AerobicTraining        kinase             2  \n"
     ]
    }
   ],
   "source": [
    "df_2['enzyme_class'] = enzyme_class\n",
    "\n",
    "print (df_2.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "8d1a7a84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "enzyme_class\n",
       "0    202\n",
       "1      8\n",
       "2     13\n",
       "3     23\n",
       "4      3\n",
       "6      4\n",
       "7      1\n",
       "Name: membership, dtype: int64"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_2.groupby(['enzyme_class'])['membership'].count()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4921d604",
   "metadata": {},
   "source": [
    "3.) The third step is the enzyme classification association of each vitamin and mineral based on their web scrapes from wikipedia and journals.  The logic is this: find a summary of each vitamin/nutrient, iterate over the text in search of the enzyme suffix, 'ase', sort out words that are clearly not enzymes, use enzyme classification catalogue to find how many times a certain enzyme classification is mentioned in the description, and then create a dictionary of percentages of associations.  From there the associations factors will be added to the dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "4676896a",
   "metadata": {},
   "outputs": [],
   "source": [
    "nutrients= ['Vitamin A','Vitamin C', 'Calcium', 'Iron', 'Vitamin D', 'Vitamin E', 'Vitamin K', 'Thiamin', 'Riboflavin', 'Niacin', 'Vitamin B6', 'Folate', 'Vitamin B12', 'Biotin', 'Pantothenic Acid', 'Phosphorous', 'Iodine', 'Magnesium', 'Zinc', 'Selenium', 'Copper', 'Manganese', 'Chromium', 'Molybdenum', 'Chloride', 'Potassium', 'Choline']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "65c663a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import wikipedia"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9c712b9",
   "metadata": {},
   "source": [
    "Below is the finalized set of words to set of enzyme classification identifications from all of the scrapes (had to personally compile this list through 27 searches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "6744d121",
   "metadata": {},
   "outputs": [],
   "source": [
    "oxidoreductase = ['oxidase', 'dehydrogenase', 'desaturase', 'peroxidase', '4-hydroxylase', 'reductase', 'dehydrogenase/reductase','dehydrogenases','reductasesretinal','dehydrogenaseretinal','1515dioxygenase','1515monooxygenase','15monooxygenase','prolyl3hydroxylases','prolyl4hydroxylases','hydroxylases','hydroxylase','dioxygenase', 'betahydroxylase', 'monooxygenase', 'oxidases','6dehydrogenase','25hydroxylase', '1alphahydroxylase', '24hydroxylase', '27hydroxylase','hydrogenases','nitrogenase','lipoxygenases','catalase','deiodinases','deiodinase','peroxidases','Deiodinases','iodinase','dehydrogenasebranched', 'dehydrogenase2','laccases','tyrosinases','reductaseChemical','oxidoreductases','phenoloxidase', 'nitrogenases','Nitrogenases','oxidoreductase']\n",
    "transferase = ['kinase','6-beta-N-acetylglucosaminyltransferase','glucosyltransferase', 'isopentenyltransferase', 'sulfotransferase','acyltransferase','transferase','pyrophosphorylase','glucuronokinase','carboxylase','transaminase','Transaminases','phosphorylase','transaminases','aminotransferase','vitaminMethyltransferasesMethyl','methyltransferase','diphosphokinase','transketolase','kinaseThe','thiaminases', 'thiaminase','kinase;','polymerase','polymerases','hydroxymethyltransferase','carboxylases','transferaseFinally','transferases','kinases','cytidylyltransferases','acetyltransferase','Kinases','carboxykinase','[53]Phosphofructokinase']\n",
    "hydrolase = ['deaminase', 'GTPase', 'GTPase,', 'ATPase', 'ATPases' 'metallopeptidase', 'caspase', 'phosphatase', 'beta-secretase', 'glutaminase', 'endonuclease', 'dihydropyrimidinase', 'peptidase', 'helicase', 'alpha-L-fucosidase', 'carboxypeptidase', 'phosphodiesterase','hydrolases','hydrolase','gluconolactonasebetween','phosphatases','vitaminDehalogenasesSome', 'dehalogenases','proteases','glucosidase','pyrophosphatase','conjugase','biotinidase','Biotinidase','sulfoxideBiotinidase','Pantetheinase','fermentationCarboxypeptidase','arginase','transcriptases','amylase','cholinephosphotransferase','phospholipases','acetylcholinesterase','Phosphatases','Enolase','helicases','exonucleases']\n",
    "lyase = ['cyclase', 'lyase', 'anhydrase','cystathionase','lyasetransketolasePresent','decarboxylase','hydratase','decarboxylase4','lyases','Aldolase']\n",
    "isomerase = ['racemase','animalsIsomerasesRearrangements','isomerase','isomerases','mutase','dismutase','dismutases','topoisomerases']\n",
    "ligase = ['synthetase', 'synthase','cosynthetase','synthaseThe','synthase;','ligase','synthetaseValproic','holocarboxylase','synthetaseBiotinylation','Holocarboxylase','ligases']\n",
    "translocase = ['scramblase']\n",
    "\n",
    "ec = [oxidoreductase, transferase, hydrolase, lyase, isomerase, ligase, translocase]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5f75c2d",
   "metadata": {},
   "source": [
    "Below is the function used to scrape the wikis for enzymes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "dd94fe71",
   "metadata": {},
   "outputs": [],
   "source": [
    "def nutrient_text_enzymes(nutrient):\n",
    "    \n",
    "    wiki = wikipedia.page(nutrient)\n",
    "\n",
    "    text = wiki.content\n",
    "\n",
    "    text = text.replace('==', '')\n",
    "\n",
    "    text = text.replace('\\n', '')[:-12]\n",
    "    \n",
    "    #print (text)\n",
    "    \n",
    "    text = text.replace(\".\", \"\")\n",
    "    text = text.replace(':','')\n",
    "    text = text.replace(',','')\n",
    "    text = text.replace('-',' ')\n",
    "    text = text.replace(\"'\",'')\n",
    "    text = text.replace('(','')\n",
    "    text = text.replace(')','')\n",
    "    text = text.replace('=','')\n",
    "    text = text.replace('\"','')\n",
    "\n",
    "\n",
    "    text = str(text)\n",
    "    text = text.split()\n",
    "    enzymes = []\n",
    "    \n",
    "    not_useful_ase = ['case', 'cases', 'increase', 'increases', 'increased', 'increasing','decrease', 'decreases', 'decreased','decreasing','release','releases','released','releasing','cease','ceases','ceased','ceasing','base','bases','based','disease','diseases','database','Cases','cerealbased','baseline','proteinbased','plantbased','increasesThe','grease','greases','greased','D3based', 'Disease', 'Diseases', 'Diseased', 'Increase','Increases','Increased','Decrease','Decreased','Decreases','Database','please','pleases','pleased''Please','Pleases','Pleased','diseases.','disease:','disease.','Disease:','Disease.']\n",
    "    \n",
    "    for x in text:\n",
    "        ase = 'ase'\n",
    "        if x not in not_useful_ase and ase in x:\n",
    "            enzymes.append(x)\n",
    "            \n",
    "    return enzymes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb62cfdb",
   "metadata": {},
   "source": [
    "Below is the enzyme classification function used in the scraping of sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "9f56db1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def enzyme_classify(enzyme_list):\n",
    "    \n",
    "    vit_distribution= []\n",
    "\n",
    "    for x in range(len(enzyme_list)):\n",
    "        if enzyme_list[x] in ec[0]:\n",
    "            vit_distribution.append(1)\n",
    "        elif enzyme_list[x] in ec[1]:\n",
    "            vit_distribution.append(2)\n",
    "        elif enzyme_list[x] in ec[2]:\n",
    "            vit_distribution.append(3)\n",
    "        elif enzyme_list[x] in ec[3]:\n",
    "            vit_distribution.append(4)\n",
    "        elif enzyme_list[x] in ec[4]:\n",
    "            vit_distribution.append(5)\n",
    "        elif enzyme_list[x] in ec[5]:\n",
    "            vit_distribution.append(6)\n",
    "        elif enzyme_list[x] in ec[6]:\n",
    "            vit_distribution.append(7)\n",
    "        else:\n",
    "            vit_distribution.append(0)\n",
    "            \n",
    "    return vit_distribution\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b41de33a",
   "metadata": {},
   "source": [
    "Below is the function used to make the classification percentages:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "f2baffba",
   "metadata": {},
   "outputs": [],
   "source": [
    "def vit_rat_dict(vit_list):\n",
    "    \n",
    "    history= []\n",
    "    \n",
    "    dict_rat ={}\n",
    "    \n",
    "    for x in range(len(vit_list)):\n",
    "        if vit_list[x] not in history and vit_list[x] != 0:\n",
    "            y = (vit_list.count(vit_list[x]))/len([i for i in vit_list if i !=0])\n",
    "            z = vit_list[x]\n",
    "            dict_rat[z]= y\n",
    "            history.append(list[x])\n",
    "        elif list[x] in history:\n",
    "            print('no')\n",
    "            \n",
    "    return dict_rat"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "016f7abd",
   "metadata": {},
   "source": [
    "Below is the function combining the: 1.)wiki search and scrape, 2.)enzyme classification, and ratio 3.)production all into one"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "df83fdc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def nutrient_overall(nutrient):\n",
    "    nutr_enzymes = nutrient_text_enzymes(nutrient)\n",
    "\n",
    "    #print (nutr_enzymes)\n",
    "    #print (len(nutr_enzymes))\n",
    "    \n",
    "    nutr_distribution = enzyme_classify(nutr_enzymes)\n",
    "\n",
    "    #print (nutr_distribution)\n",
    "\n",
    "    ratio_nutr = vit_rat_dict(nutr_distribution)\n",
    "\n",
    "    #print (ratio_nutr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1e047c1",
   "metadata": {},
   "source": [
    "Below is the function used to make scrapes on journal entries using Beatiful Soup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "933493b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "from bs4 import BeautifulSoup\n",
    "import pandas as pd\n",
    "import time"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8a866c4",
   "metadata": {},
   "source": [
    "This is the function to scrape the url:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "ea0f183c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def ncbi_enzyme_search(url):\n",
    "\n",
    "\n",
    "    response = requests.get(url)\n",
    "\n",
    "    soup = BeautifulSoup(response.content, 'html.parser')\n",
    "    text = soup.get_text()\n",
    "    \n",
    "    \n",
    "    text = text.replace(\".\", \"\")\n",
    "    text = text.replace(':','')\n",
    "    text = text.replace(',','')\n",
    "    text = text.replace('-',' ')\n",
    "    text = text.replace(\"'\",'')\n",
    "    text = text.replace('(','')\n",
    "    text = text.replace(')','')\n",
    "    text = text.replace('=','')\n",
    "    text = text.replace('\"','')\n",
    "    \n",
    "    text = text.split()\n",
    "    enzymes = []\n",
    "    \n",
    "    not_useful_ase = ['case', 'cases', 'increase', 'increases', 'increased', 'increasing','decrease', 'decreases', 'decreased','decreasing','release','releases','released','releasing','cease','ceases','ceased','ceasing','base','bases','based','disease','diseases','database','Cases','cerealbased','baseline','proteinbased','plantbased','increasesThe','grease','greases','greased','D3based', 'Disease', 'Diseases', 'Diseased', 'Increase','Increases','Increased','Decrease','Decreased','Decreases','Database','please','pleases','pleased''Please','Pleases','Pleased','diseases.','disease:','disease.','Disease:','Disease.']\n",
    "\n",
    "    for x in text:\n",
    "        ase = 'ase'\n",
    "        if x not in not_useful_ase and ase in x:\n",
    "            enzymes.append(x)\n",
    "            \n",
    "    return enzymes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c94cab47",
   "metadata": {},
   "source": [
    "This is the function to produce the dictionary of association percentages (utilizing the functions used in the wiki scrapes):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "668afdc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def nutrient_overall_url(url):\n",
    "    nutr_enzymes = ncbi_enzyme_search(url)\n",
    "\n",
    "    #print (nutr_enzymes)\n",
    "    #print (len(nutr_enzymes))\n",
    "    \n",
    "    nutr_distribution = enzyme_classify(nutr_enzymes)\n",
    "\n",
    "    #print (nutr_distribution)\n",
    "\n",
    "    ratio_nutr = vit_rat_dict(nutr_distribution)\n",
    "\n",
    "    #print (ratio_nutr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "179aa2c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def nutrient_dataframe(nutrient_list):\n",
    "    df_list = []\n",
    "    \n",
    "    for x in nutrient_list:\n",
    "    \n",
    "        nutr_enzymes = nutrient_text_enzymes(x)\n",
    "    \n",
    "        nutr_distribution = enzyme_classify(nutr_enzymes)\n",
    "\n",
    "        ratio_nutr = vit_rat_dict(nutr_distribution)\n",
    "    \n",
    "        df_list.append(ratio_nutr)\n",
    "    \n",
    "    df_nutr = pd.DataFrame(data=df_list)\n",
    "    \n",
    "    return df_nutr\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "b6fd85f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "nutrient_list_wiki = ['Vitamin A','Vitamin.C','Iron.','Vitamin.D','Vitamin.E','Vitamin.K','Thiamin.','Riboflavin.','Niacin.','Vitamin.B6','Folate.','Vitamin.B12','Biotin.','Pantothenic Acid.','Iodine.','Zinc.','Selenium','coopper','Manganese','Molybdenum','Chloride','Potassium','Choline.']\n",
    "\n",
    "df_nutr = nutrient_dataframe(nutrient_list_wiki)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "0442591e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>20</th>\n",
       "      <th>21</th>\n",
       "      <th>22</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.85</td>\n",
       "      <td>0.78125</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.260870</td>\n",
       "      <td>0.470588</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.217391</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.133333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.10</td>\n",
       "      <td>0.06250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>0.117647</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>...</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.045455</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.05</td>\n",
       "      <td>0.09375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.478261</td>\n",
       "      <td>0.117647</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.391304</td>\n",
       "      <td>...</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.466667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.06250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.130435</td>\n",
       "      <td>0.294118</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.130435</td>\n",
       "      <td>...</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.130435</td>\n",
       "      <td>...</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.444444</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.066667</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     0        1    2         3    4         5         6         7         8   \\\n",
       "1  0.85  0.78125  1.0  0.857143  0.5  0.166667  0.260870  0.470588  0.142857   \n",
       "3  0.10  0.06250  NaN       NaN  NaN       NaN  0.043478  0.117647       NaN   \n",
       "2  0.05  0.09375  NaN       NaN  0.5  0.833333  0.478261  0.117647  0.714286   \n",
       "6   NaN  0.06250  NaN  0.142857  NaN       NaN  0.130435  0.294118       NaN   \n",
       "4   NaN      NaN  NaN       NaN  NaN       NaN  0.086957       NaN  0.142857   \n",
       "\n",
       "         9   ...        13   14        15        16        17        18   19  \\\n",
       "1  0.217391  ...       NaN  1.0  0.222222  0.909091  0.777778  0.230769  1.0   \n",
       "3  0.086957  ...  0.142857  NaN  0.333333  0.045455       NaN  0.230769  NaN   \n",
       "2  0.391304  ...  0.571429  NaN       NaN       NaN       NaN  0.076923  NaN   \n",
       "6  0.130435  ...  0.142857  NaN       NaN       NaN       NaN  0.230769  NaN   \n",
       "4  0.130435  ...  0.142857  NaN  0.444444       NaN       NaN  0.076923  NaN   \n",
       "\n",
       "    20   21        22  \n",
       "1  NaN  NaN  0.133333  \n",
       "3  1.0  1.0  0.333333  \n",
       "2  NaN  NaN  0.466667  \n",
       "6  NaN  NaN       NaN  \n",
       "4  NaN  NaN  0.066667  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_nutr_T =df_nutr.T\n",
    "df_nutr_T.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "aeabd8e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_nutr_T.sort_index(axis=0, ascending=True, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "0c9661e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "nutrients= ['Vitamin A Factor','Vitamin C Factor', 'Iron Factor', 'Vitamin D Factor', 'Vitamin E Factor', 'Vitamin K Factor', 'Thiamin Factor', 'Riboflavin Factor', 'Niacin Factor', 'Vitamin B6 Factor', 'Folate Factor', 'Vitamin B12 Factor', 'Biotin Factor', 'Pantothenic Acid Factor', 'Iodine Factor', 'Zinc Factor', 'Selenium Factor', 'Copper Factor', 'Manganese Factor', 'Molybdenum Factor', 'Chloride Factor', 'Potassium Factor', 'Choline Factor']\n",
    "\n",
    "df_nutr_T.columns = nutrients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "40525b69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Vitamin A Factor</th>\n",
       "      <th>Vitamin C Factor</th>\n",
       "      <th>Iron Factor</th>\n",
       "      <th>Vitamin D Factor</th>\n",
       "      <th>Vitamin E Factor</th>\n",
       "      <th>Vitamin K Factor</th>\n",
       "      <th>Thiamin Factor</th>\n",
       "      <th>Riboflavin Factor</th>\n",
       "      <th>Niacin Factor</th>\n",
       "      <th>Vitamin B6 Factor</th>\n",
       "      <th>...</th>\n",
       "      <th>Iodine Factor</th>\n",
       "      <th>Zinc Factor</th>\n",
       "      <th>Selenium Factor</th>\n",
       "      <th>Copper Factor</th>\n",
       "      <th>Manganese Factor</th>\n",
       "      <th>Molybdenum Factor</th>\n",
       "      <th>Chloride Factor</th>\n",
       "      <th>Potassium Factor</th>\n",
       "      <th>Choline Factor</th>\n",
       "      <th>enzyme_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.85</td>\n",
       "      <td>0.78125</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.260870</td>\n",
       "      <td>0.470588</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.217391</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.05</td>\n",
       "      <td>0.09375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.478261</td>\n",
       "      <td>0.117647</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.391304</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.466667</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.10</td>\n",
       "      <td>0.06250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>0.117647</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.045455</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.130435</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.444444</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.066667</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.045455</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Vitamin A Factor  Vitamin C Factor  Iron Factor  Vitamin D Factor  \\\n",
       "1              0.85           0.78125          1.0          0.857143   \n",
       "2              0.05           0.09375          NaN               NaN   \n",
       "3              0.10           0.06250          NaN               NaN   \n",
       "4               NaN               NaN          NaN               NaN   \n",
       "5               NaN               NaN          NaN               NaN   \n",
       "\n",
       "   Vitamin E Factor  Vitamin K Factor  Thiamin Factor  Riboflavin Factor  \\\n",
       "1               0.5          0.166667        0.260870           0.470588   \n",
       "2               0.5          0.833333        0.478261           0.117647   \n",
       "3               NaN               NaN        0.043478           0.117647   \n",
       "4               NaN               NaN        0.086957                NaN   \n",
       "5               NaN               NaN             NaN                NaN   \n",
       "\n",
       "   Niacin Factor  Vitamin B6 Factor  ...  Iodine Factor  Zinc Factor  \\\n",
       "1       0.142857           0.217391  ...            1.0     0.222222   \n",
       "2       0.714286           0.391304  ...            NaN          NaN   \n",
       "3            NaN           0.086957  ...            NaN     0.333333   \n",
       "4       0.142857           0.130435  ...            NaN     0.444444   \n",
       "5            NaN           0.043478  ...            NaN          NaN   \n",
       "\n",
       "   Selenium Factor  Copper Factor  Manganese Factor  Molybdenum Factor  \\\n",
       "1         0.909091       0.777778          0.230769                1.0   \n",
       "2              NaN            NaN          0.076923                NaN   \n",
       "3         0.045455            NaN          0.230769                NaN   \n",
       "4              NaN            NaN          0.076923                NaN   \n",
       "5         0.045455       0.222222          0.153846                NaN   \n",
       "\n",
       "   Chloride Factor  Potassium Factor  Choline Factor  enzyme_class  \n",
       "1              NaN               NaN        0.133333             1  \n",
       "2              NaN               NaN        0.466667             2  \n",
       "3              1.0               1.0        0.333333             3  \n",
       "4              NaN               NaN        0.066667             4  \n",
       "5              NaN               NaN             NaN             5  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_nutr_T['enzyme_class']= {1:1, 2:2, 3:3, 4:4, 5:5, 6:6, 7:7}\n",
    "df_nutr_T.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fbb00f0",
   "metadata": {},
   "source": [
    "Do a second nutrient dataframe addition for the journal entries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "9b9b8e34",
   "metadata": {},
   "outputs": [],
   "source": [
    "def nutrient_dataframe_sec(url_list):\n",
    "    df_list = []\n",
    "    \n",
    "    for x in url_list:\n",
    "    \n",
    "        nutr_enzymes = ncbi_enzyme_search(x)\n",
    "    \n",
    "        nutr_distribution = enzyme_classify(nutr_enzymes)\n",
    "\n",
    "        ratio_nutr = vit_rat_dict(nutr_distribution)\n",
    "    \n",
    "        df_list.append(ratio_nutr)\n",
    "    \n",
    "    df_nutr = pd.DataFrame(data=df_list)\n",
    "    \n",
    "    return df_nutr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "5abe871f",
   "metadata": {},
   "outputs": [],
   "source": [
    "url_list_nutrients = ['https://www.mdpi.com/2072-6643/13/4/1136', 'https://www.mdpi.com/2072-6643/13/4/1136', 'https://www.mdpi.com/2218-273X/12/12/1891', 'https://www.mdpi.com/1422-0067/20/19/4901']\n",
    "\n",
    "df_nutr_2 = nutrient_dataframe_sec(url_list_nutrients)\n",
    "\n",
    "df_nutr_2_T =df_nutr_2.T\n",
    "df_nutr_2_T.head()\n",
    "\n",
    "df_nutr_2_T.sort_index(axis=0, ascending=True, inplace=True)\n",
    "\n",
    "nutrient_sec = ['Magnesium Factor', 'Phosphate Factor', 'Calcium Factor', 'Chromium Factor']\n",
    "df_nutr_2_T.columns = nutrient_sec\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "dd94e343",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Magnesium Factor</th>\n",
       "      <th>Phosphate Factor</th>\n",
       "      <th>Calcium Factor</th>\n",
       "      <th>Chromium Factor</th>\n",
       "      <th>enzyme_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.228571</td>\n",
       "      <td>0.228571</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.538462</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.300000</td>\n",
       "      <td>0.300000</td>\n",
       "      <td>0.307692</td>\n",
       "      <td>0.384615</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.014286</td>\n",
       "      <td>0.014286</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.028571</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Magnesium Factor  Phosphate Factor  Calcium Factor  Chromium Factor  \\\n",
       "1          0.228571          0.228571        0.230769         0.076923   \n",
       "2          0.400000          0.400000        0.153846         0.538462   \n",
       "3          0.300000          0.300000        0.307692         0.384615   \n",
       "4          0.014286          0.014286             NaN              NaN   \n",
       "5          0.028571          0.028571             NaN              NaN   \n",
       "\n",
       "   enzyme_class  \n",
       "1             1  \n",
       "2             2  \n",
       "3             3  \n",
       "4             4  \n",
       "5             5  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_nutr_2_T['enzyme_class']= {1:1, 2:2, 3:3, 4:4, 5:5, 6:6, 7:7}\n",
    "df_nutr_2_T.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "2206544b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Vitamin A Factor</th>\n",
       "      <th>Vitamin C Factor</th>\n",
       "      <th>Iron Factor</th>\n",
       "      <th>Vitamin D Factor</th>\n",
       "      <th>Vitamin E Factor</th>\n",
       "      <th>Vitamin K Factor</th>\n",
       "      <th>Thiamin Factor</th>\n",
       "      <th>Riboflavin Factor</th>\n",
       "      <th>Niacin Factor</th>\n",
       "      <th>Vitamin B6 Factor</th>\n",
       "      <th>...</th>\n",
       "      <th>Manganese Factor</th>\n",
       "      <th>Molybdenum Factor</th>\n",
       "      <th>Chloride Factor</th>\n",
       "      <th>Potassium Factor</th>\n",
       "      <th>Choline Factor</th>\n",
       "      <th>enzyme_class</th>\n",
       "      <th>Magnesium Factor</th>\n",
       "      <th>Phosphate Factor</th>\n",
       "      <th>Calcium Factor</th>\n",
       "      <th>Chromium Factor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.85</td>\n",
       "      <td>0.78125</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.260870</td>\n",
       "      <td>0.470588</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.217391</td>\n",
       "      <td>...</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>1</td>\n",
       "      <td>0.228571</td>\n",
       "      <td>0.228571</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.076923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.05</td>\n",
       "      <td>0.09375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.478261</td>\n",
       "      <td>0.117647</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.391304</td>\n",
       "      <td>...</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.466667</td>\n",
       "      <td>2</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.538462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.10</td>\n",
       "      <td>0.06250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>0.117647</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>...</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>3</td>\n",
       "      <td>0.300000</td>\n",
       "      <td>0.300000</td>\n",
       "      <td>0.307692</td>\n",
       "      <td>0.384615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.130435</td>\n",
       "      <td>...</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.066667</td>\n",
       "      <td>4</td>\n",
       "      <td>0.014286</td>\n",
       "      <td>0.014286</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>...</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Vitamin A Factor  Vitamin C Factor  Iron Factor  Vitamin D Factor  \\\n",
       "0              0.85           0.78125          1.0          0.857143   \n",
       "1              0.05           0.09375          NaN               NaN   \n",
       "2              0.10           0.06250          NaN               NaN   \n",
       "3               NaN               NaN          NaN               NaN   \n",
       "4               NaN               NaN          NaN               NaN   \n",
       "\n",
       "   Vitamin E Factor  Vitamin K Factor  Thiamin Factor  Riboflavin Factor  \\\n",
       "0               0.5          0.166667        0.260870           0.470588   \n",
       "1               0.5          0.833333        0.478261           0.117647   \n",
       "2               NaN               NaN        0.043478           0.117647   \n",
       "3               NaN               NaN        0.086957                NaN   \n",
       "4               NaN               NaN             NaN                NaN   \n",
       "\n",
       "   Niacin Factor  Vitamin B6 Factor  ...  Manganese Factor  Molybdenum Factor  \\\n",
       "0       0.142857           0.217391  ...          0.230769                1.0   \n",
       "1       0.714286           0.391304  ...          0.076923                NaN   \n",
       "2            NaN           0.086957  ...          0.230769                NaN   \n",
       "3       0.142857           0.130435  ...          0.076923                NaN   \n",
       "4            NaN           0.043478  ...          0.153846                NaN   \n",
       "\n",
       "   Chloride Factor  Potassium Factor  Choline Factor  enzyme_class  \\\n",
       "0              NaN               NaN        0.133333             1   \n",
       "1              NaN               NaN        0.466667             2   \n",
       "2              1.0               1.0        0.333333             3   \n",
       "3              NaN               NaN        0.066667             4   \n",
       "4              NaN               NaN             NaN             5   \n",
       "\n",
       "   Magnesium Factor  Phosphate Factor  Calcium Factor  Chromium Factor  \n",
       "0          0.228571          0.228571        0.230769         0.076923  \n",
       "1          0.400000          0.400000        0.153846         0.538462  \n",
       "2          0.300000          0.300000        0.307692         0.384615  \n",
       "3          0.014286          0.014286             NaN              NaN  \n",
       "4          0.028571          0.028571             NaN              NaN  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_nutr_merge = df_nutr_T.merge(df_nutr_2_T, left_on='enzyme_class', right_on='enzyme_class', how='left')\n",
    "\n",
    "df_nutr_merge.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16b79f3e",
   "metadata": {},
   "source": [
    "Now merge the nutrient dataframes to the original"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "36887a3d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   membership  traitsig        ensg_version      gene LV_number  \\\n",
      "0      0.5158    0.2172  ENSG00000137959.15    IFI44L         2   \n",
      "1      0.5497    0.4713  ENSG00000160710.16      ADAR         2   \n",
      "2      0.7754    0.5447  ENSG00000082512.14     TRAF5         2   \n",
      "3      0.7309    0.4325  ENSG00000163219.11  ARHGAP25         2   \n",
      "4      0.5587    0.4396  ENSG00000153208.16     MERTK         2   \n",
      "5      0.6029    0.5140   ENSG00000152127.8     MGAT5         2   \n",
      "6      0.4493    0.3355  ENSG00000163359.15    COL6A3         2   \n",
      "7      0.8284    0.6855  ENSG00000168016.14    TRANK1         2   \n",
      "8      0.7394    0.5133  ENSG00000188313.12    PLSCR1         2   \n",
      "9      0.6431    0.7271   ENSG00000114450.9      GNB4         2   \n",
      "\n",
      "                LV_pathway                           assoc_enzymes  \\\n",
      "0  MetaMEx_AerobicTraining                                   other   \n",
      "1  MetaMEx_AerobicTraining                               deaminase   \n",
      "2  MetaMEx_AerobicTraining                                   other   \n",
      "3  MetaMEx_AerobicTraining                                  GTPase   \n",
      "4  MetaMEx_AerobicTraining                                  kinase   \n",
      "5  MetaMEx_AerobicTraining  6-beta-N-acetylglucosaminyltransferase   \n",
      "6  MetaMEx_AerobicTraining                                   other   \n",
      "7  MetaMEx_AerobicTraining                                   other   \n",
      "8  MetaMEx_AerobicTraining                              scramblase   \n",
      "9  MetaMEx_AerobicTraining                                   other   \n",
      "\n",
      "   enzyme_class  Vitamin A Factor  Vitamin C Factor  ...  Copper Factor  \\\n",
      "0             0               NaN               NaN  ...            NaN   \n",
      "1             3              0.10           0.06250  ...            NaN   \n",
      "2             0               NaN               NaN  ...            NaN   \n",
      "3             3              0.10           0.06250  ...            NaN   \n",
      "4             2              0.05           0.09375  ...            NaN   \n",
      "5             2              0.05           0.09375  ...            NaN   \n",
      "6             0               NaN               NaN  ...            NaN   \n",
      "7             0               NaN               NaN  ...            NaN   \n",
      "8             7               NaN               NaN  ...            NaN   \n",
      "9             0               NaN               NaN  ...            NaN   \n",
      "\n",
      "   Manganese Factor  Molybdenum Factor  Chloride Factor  Potassium Factor  \\\n",
      "0               NaN                NaN              NaN               NaN   \n",
      "1          0.230769                NaN              1.0               1.0   \n",
      "2               NaN                NaN              NaN               NaN   \n",
      "3          0.230769                NaN              1.0               1.0   \n",
      "4          0.076923                NaN              NaN               NaN   \n",
      "5          0.076923                NaN              NaN               NaN   \n",
      "6               NaN                NaN              NaN               NaN   \n",
      "7               NaN                NaN              NaN               NaN   \n",
      "8               NaN                NaN              NaN               NaN   \n",
      "9               NaN                NaN              NaN               NaN   \n",
      "\n",
      "   Choline Factor  Magnesium Factor  Phosphate Factor  Calcium Factor  \\\n",
      "0             NaN               NaN               NaN             NaN   \n",
      "1        0.333333               0.3               0.3        0.307692   \n",
      "2             NaN               NaN               NaN             NaN   \n",
      "3        0.333333               0.3               0.3        0.307692   \n",
      "4        0.466667               0.4               0.4        0.153846   \n",
      "5        0.466667               0.4               0.4        0.153846   \n",
      "6             NaN               NaN               NaN             NaN   \n",
      "7             NaN               NaN               NaN             NaN   \n",
      "8             NaN               NaN               NaN             NaN   \n",
      "9             NaN               NaN               NaN             NaN   \n",
      "\n",
      "   Chromium Factor  \n",
      "0              NaN  \n",
      "1         0.384615  \n",
      "2              NaN  \n",
      "3         0.384615  \n",
      "4         0.538462  \n",
      "5         0.538462  \n",
      "6              NaN  \n",
      "7              NaN  \n",
      "8              NaN  \n",
      "9              NaN  \n",
      "\n",
      "[10 rows x 35 columns]\n"
     ]
    }
   ],
   "source": [
    "df_merge = df_2.merge(df_nutr_merge, left_on='enzyme_class', right_on='enzyme_class', how='left')\n",
    "\n",
    "print (df_merge.head(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "7b2167a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merge = df_merge.drop(columns=['LV_number','LV_pathway'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "0846d5e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merge = df_merge.fillna(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdbd99f4",
   "metadata": {},
   "source": [
    "Make the multiplication columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "6ab779a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "vit_subset = df_merge.iloc[: ,6:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "5ff63c83",
   "metadata": {},
   "outputs": [],
   "source": [
    "def assoc_products(subset):\n",
    "    data_frame = pd.DataFrame()\n",
    "    for x in range(len(subset.columns)):\n",
    "        coll = df_merge['membership'] * vit_subset[subset.columns[x]] + df_merge['traitsig'] * vit_subset[subset.columns[x]]\n",
    "        \n",
    "        ser = pd.Series(data = coll)\n",
    "        \n",
    "        data_frame[subset.columns[x]+' Product and Sum'] = ser\n",
    "        \n",
    "    return data_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "1884ac4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Vitamin A Factor Product and Sum  Vitamin C Factor Product and Sum  \\\n",
      "0                            0.000000                          0.000000   \n",
      "1                            0.102100                          0.063812   \n",
      "2                            0.000000                          0.000000   \n",
      "3                            0.116340                          0.072712   \n",
      "4                            0.049915                          0.093591   \n",
      "..                                ...                               ...   \n",
      "249                          0.000000                          0.000000   \n",
      "250                          0.000000                          0.000000   \n",
      "251                          0.000000                          0.000000   \n",
      "252                          0.038750                          0.072656   \n",
      "253                          0.102260                          0.063912   \n",
      "\n",
      "     Iron Factor Product and Sum  Vitamin D Factor Product and Sum  \\\n",
      "0                            0.0                               0.0   \n",
      "1                            0.0                               0.0   \n",
      "2                            0.0                               0.0   \n",
      "3                            0.0                               0.0   \n",
      "4                            0.0                               0.0   \n",
      "..                           ...                               ...   \n",
      "249                          0.0                               0.0   \n",
      "250                          0.0                               0.0   \n",
      "251                          0.0                               0.0   \n",
      "252                          0.0                               0.0   \n",
      "253                          0.0                               0.0   \n",
      "\n",
      "     Vitamin E Factor Product and Sum  Vitamin K Factor Product and Sum  \\\n",
      "0                             0.00000                          0.000000   \n",
      "1                             0.00000                          0.000000   \n",
      "2                             0.00000                          0.000000   \n",
      "3                             0.00000                          0.000000   \n",
      "4                             0.49915                          0.831917   \n",
      "..                                ...                               ...   \n",
      "249                           0.00000                          0.000000   \n",
      "250                           0.00000                          0.000000   \n",
      "251                           0.00000                          0.000000   \n",
      "252                           0.38750                          0.645833   \n",
      "253                           0.00000                          0.000000   \n",
      "\n",
      "     Thiamin Factor Product and Sum  Riboflavin Factor Product and Sum  \\\n",
      "0                          0.000000                           0.000000   \n",
      "1                          0.044391                           0.120118   \n",
      "2                          0.000000                           0.000000   \n",
      "3                          0.050583                           0.136871   \n",
      "4                          0.477448                           0.117447   \n",
      "..                              ...                                ...   \n",
      "249                        0.000000                           0.000000   \n",
      "250                        0.000000                           0.000000   \n",
      "251                        0.000000                           0.000000   \n",
      "252                        0.370652                           0.091176   \n",
      "253                        0.044461                           0.120306   \n",
      "\n",
      "     Niacin Factor Product and Sum  Vitamin B6 Factor Product and Sum  ...  \\\n",
      "0                         0.000000                           0.000000  ...   \n",
      "1                         0.000000                           0.088783  ...   \n",
      "2                         0.000000                           0.000000  ...   \n",
      "3                         0.000000                           0.101165  ...   \n",
      "4                         0.713071                           0.390639  ...   \n",
      "..                             ...                                ...  ...   \n",
      "249                       0.000000                           0.000000  ...   \n",
      "250                       0.000000                           0.000000  ...   \n",
      "251                       0.000000                           0.000000  ...   \n",
      "252                       0.553571                           0.303261  ...   \n",
      "253                       0.000000                           0.088922  ...   \n",
      "\n",
      "     Copper Factor Product and Sum  Manganese Factor Product and Sum  \\\n",
      "0                              0.0                          0.000000   \n",
      "1                              0.0                          0.235615   \n",
      "2                              0.0                          0.000000   \n",
      "3                              0.0                          0.268477   \n",
      "4                              0.0                          0.076792   \n",
      "..                             ...                               ...   \n",
      "249                            0.0                          0.000000   \n",
      "250                            0.0                          0.000000   \n",
      "251                            0.0                          0.000000   \n",
      "252                            0.0                          0.059615   \n",
      "253                            0.0                          0.235985   \n",
      "\n",
      "     Molybdenum Factor Product and Sum  Chloride Factor Product and Sum  \\\n",
      "0                                  0.0                           0.0000   \n",
      "1                                  0.0                           1.0210   \n",
      "2                                  0.0                           0.0000   \n",
      "3                                  0.0                           1.1634   \n",
      "4                                  0.0                           0.0000   \n",
      "..                                 ...                              ...   \n",
      "249                                0.0                           0.0000   \n",
      "250                                0.0                           0.0000   \n",
      "251                                0.0                           0.0000   \n",
      "252                                0.0                           0.0000   \n",
      "253                                0.0                           1.0226   \n",
      "\n",
      "     Potassium Factor Product and Sum  Choline Factor Product and Sum  \\\n",
      "0                              0.0000                        0.000000   \n",
      "1                              1.0210                        0.340333   \n",
      "2                              0.0000                        0.000000   \n",
      "3                              1.1634                        0.387800   \n",
      "4                              0.0000                        0.465873   \n",
      "..                                ...                             ...   \n",
      "249                            0.0000                        0.000000   \n",
      "250                            0.0000                        0.000000   \n",
      "251                            0.0000                        0.000000   \n",
      "252                            0.0000                        0.361667   \n",
      "253                            1.0226                        0.340867   \n",
      "\n",
      "     Magnesium Factor Product and Sum  Phosphate Factor Product and Sum  \\\n",
      "0                             0.00000                           0.00000   \n",
      "1                             0.30630                           0.30630   \n",
      "2                             0.00000                           0.00000   \n",
      "3                             0.34902                           0.34902   \n",
      "4                             0.39932                           0.39932   \n",
      "..                                ...                               ...   \n",
      "249                           0.00000                           0.00000   \n",
      "250                           0.00000                           0.00000   \n",
      "251                           0.00000                           0.00000   \n",
      "252                           0.31000                           0.31000   \n",
      "253                           0.30678                           0.30678   \n",
      "\n",
      "     Calcium Factor Product and Sum  Chromium Factor Product and Sum  \n",
      "0                          0.000000                         0.000000  \n",
      "1                          0.314154                         0.392692  \n",
      "2                          0.000000                         0.000000  \n",
      "3                          0.357969                         0.447462  \n",
      "4                          0.153585                         0.537546  \n",
      "..                              ...                              ...  \n",
      "249                        0.000000                         0.000000  \n",
      "250                        0.000000                         0.000000  \n",
      "251                        0.000000                         0.000000  \n",
      "252                        0.119231                         0.417308  \n",
      "253                        0.314646                         0.393308  \n",
      "\n",
      "[254 rows x 27 columns]\n"
     ]
    }
   ],
   "source": [
    "df_sum_prod = assoc_products(vit_subset)\n",
    "print (df_sum_prod)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "e80343d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "254\n"
     ]
    }
   ],
   "source": [
    "df_merge_final = df_merge.merge(df_sum_prod, left_index=True, right_index=True)\n",
    "\n",
    "print (len(df_merge_final))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "748cf6a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merge_final.to_csv('wrangled_data.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e328c223",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
